 include/linux/mm.h |    1 +
 1 files changed, 1 insertion(+)

--- linux-2.4.20-b_llpio-l21/include/linux/mm.h~add_page_private	2003-07-21 21:42:50.000000000 -0700
+++ linux-2.4.20-b_llpio-l21-zab/include/linux/mm.h	2003-07-21 21:44:16.000000000 -0700
@@ -162,6 +162,7 @@ typedef struct page {
 					   protected by pagemap_lru_lock !! */
 	struct page **pprev_hash;	/* Complement to *next_hash. */
 	struct buffer_head * buffers;	/* Buffer maps us to a disk block. */
+	unsigned long private;
 
 	/*
 	 * On machines where all RAM is mapped into kernel address space,

_
