Index: linux-ia64/include/linux/mm.h
===================================================================
--- linux-ia64.orig/include/linux/mm.h	2004-04-27 12:39:16.000000000 -0700
+++ linux-ia64/include/linux/mm.h	2004-04-27 12:42:10.000000000 -0700
@@ -205,6 +205,7 @@ typedef struct page {
 	struct page **pprev_hash;	/* Complement to *next_hash. */
 	struct buffer_head * buffers;	/* Buffer maps us to a disk block. */
+	unsigned long private;
 
 	/*
 	 * On machines where all RAM is mapped into kernel address space,

_
Index: linux-ia64/mm/filemap.c
===================================================================
--- linux-ia64.orig/mm/filemap.c       2004-04-27 12:35:18.000000000 -0700
+++ linux-ia64/mm/filemap.c    2004-04-27 12:42:47.000000000 -0700
@@ -676,6 +676,7 @@ static inline void __add_to_page_cache(s
 	page_cache_get(page);
 	page->index = offset;
+	page->private = 0;
 	add_page_to_inode_queue(mapping, page);
 	add_page_to_hash_queue(page, hash);
