--- linux-2.6.27.19-5.orig/include/linux/blkdev.h	2009-03-17 12:48:06.000000000 -0400
+++ linux-2.6.27.19-5/include/linux/blkdev.h	2009-09-04 18:04:43.000000000 -0400
@@ -904,7 +904,7 @@ extern void blk_set_cmd_filter_defaults(
 #define MAX_PHYS_SEGMENTS 128
 #define MAX_HW_SEGMENTS 128
 #define SAFE_MAX_SECTORS 255
-#define BLK_DEF_MAX_SECTORS 1024
+#define BLK_DEF_MAX_SECTORS 2048
 
 #define MAX_SEGMENT_SIZE	65536
 
