Index: linux-2616-46014/include/scsi/scsi_host.h
===================================================================
--- linux-2616-46014.orig/include/scsi/scsi_host.h
+++ linux-2616-46014/include/scsi/scsi_host.h
@@ -28,7 +28,7 @@ struct scsi_transport_template;
  *	 used in one scatter-gather request.
  */
 #define SG_NONE 0
-#define SG_ALL 0xff
+#define SG_ALL 256
 
 
 #define DISABLE_CLUSTERING 0
