Index: linux-2.6.18-164.11.1/include/linux/blkdev.h
===================================================================
--- linux-2.6.18-164.11.1.orig/include/linux/blkdev.h
+++ linux-2.6.18-164.11.1/include/linux/blkdev.h
@@ -788,10 +788,10 @@ extern void blk_free_tags(struct blk_que
 extern void blk_rq_bio_prep(request_queue_t *, struct request *, struct bio *);
 extern int blkdev_issue_flush(struct block_device *, sector_t *);
 
-#define MAX_PHYS_SEGMENTS 128
-#define MAX_HW_SEGMENTS 128
+#define MAX_PHYS_SEGMENTS 256
+#define MAX_HW_SEGMENTS 256
 #define SAFE_MAX_SECTORS 255
-#define BLK_DEF_MAX_SECTORS 1024
+#define BLK_DEF_MAX_SECTORS 2048
 
 #define MAX_SEGMENT_SIZE	65536
 
Index: linux-2.6.18-164.11.1/include/scsi/scsi_host.h
===================================================================
--- linux-2.6.18-164.11.1.orig/include/scsi/scsi_host.h
+++ linux-2.6.18-164.11.1/include/scsi/scsi_host.h
@@ -30,7 +30,7 @@ struct blk_queue_tags;
  *	 used in one scatter-gather request.
  */
 #define SG_NONE 0
-#define SG_ALL 0xff
+#define SG_ALL 256
 
 
 #define DISABLE_CLUSTERING 0
Index: linux-2.6.18-164.11.1/drivers/scsi/isci/init.c
===================================================================
--- linux-2.6.18-164.11.1/drivers/scsi/isci/init.c
+++ linux-2.6.18-164.11.1/drivers/scsi/isci/init.c
@@ -38,7 +38,7 @@
 	.can_queue			= ISCI_CAN_QUEUE_VAL,
 	.cmd_per_lun			= 1,
 	.this_id			= -1,
-	.sg_tablesize			= SG_ALL,
+	.sg_tablesize			= 0x80,
 	.max_sectors			= SCSI_DEFAULT_MAX_SECTORS,
 	.use_clustering			= ENABLE_CLUSTERING,
 	.eh_device_reset_handler	= sas_eh_device_reset_handler,
Index: linux-2.6.18-164.11.1/drivers/scsi/lpfc/lpfc.h
===================================================================
--- linux-2.6.18-164.11.1.orig/drivers/scsi/lpfc/lpfc.h
+++ linux-2.6.18-164.11.1/drivers/scsi/lpfc/lpfc.h
@@ -38,7 +38,7 @@
 #define LPFC_MAX_NS_RETRY	3	/* Number of retry attempts to contact
 					   the NameServer  before giving up. */
 #define LPFC_CMD_PER_LUN	3	/* max outstanding cmds per lun */
-#define LPFC_DEFAULT_SG_SEG_CNT	64	/* sg element count per scsi cmnd */
+#define LPFC_DEFAULT_SG_SEG_CNT	256	/* sg element count per scsi cmnd */
 #define LPFC_MAX_SG_SEG_CNT	256	/* sg element count per scsi cmnd */
 #define LPFC_IOCB_LIST_CNT	2250	/* list of IOCBs for fast-path usage. */
 #define LPFC_Q_RAMP_UP_INTERVAL 120     /* lun q_depth ramp up interval */
