diff -X dontdiff -urp kern_oldest/drivers/i2c/i2c-ali1535.c kern_fix/drivers/i2c/i2c-ali1535.c
--- kern_oldest/drivers/i2c/i2c-ali1535.c	2006-05-01 11:55:43.000000000 -0700
+++ kern_fix/drivers/i2c/i2c-ali1535.c	2006-05-01 15:25:34.000000000 -0700
@@ -672,8 +672,8 @@ EXPORT_NO_SYMBOLS;
 #ifdef MODULE
 
 MODULE_AUTHOR
-    ("Frodo Looijaard <frodol@dds.nl>, Philip Edelbrock <phil@netroedge.com>,
-      Mark D. Studebaker <mdsxyz123@yahoo.com> and Dan Eaton <dan.eaton@rocketlogix.com>");
+    ("Frodo Looijaard <frodol@dds.nl>, Philip Edelbrock <phil@netroedge.com>, "
+      "Mark D. Studebaker <mdsxyz123@yahoo.com> and Dan Eaton <dan.eaton@rocketlogix.com>");
 MODULE_DESCRIPTION("ALI1535 SMBus driver");
 
 int init_module(void)
diff -X dontdiff -urp kern_oldest/drivers/net/pcmcia/wvlan_cs.c kern_fix/drivers/net/pcmcia/wvlan_cs.c
--- kern_oldest/drivers/net/pcmcia/wvlan_cs.c	2006-05-01 11:55:40.000000000 -0700
+++ kern_fix/drivers/net/pcmcia/wvlan_cs.c	2006-05-01 15:25:34.000000000 -0700
@@ -1084,9 +1084,9 @@ static int wvlan_hw_config (struct net_d
  			/* This is a PrismII card. It is is *very* similar
  			 * to the Lucent, and the driver work 95%,
  			 * therefore, we attempt to support it... */
- 			printk(KERN_NOTICE "%s: This is a PrismII card, not a Wavelan IEEE card :-(
-You may want report firmare revision (0x%X) and what the card support.
-I will try to make it work, but you should look for a better driver.\n", dev_info, firmware);
+ 			printk(KERN_NOTICE "%s: This is a PrismII card, not a Wavelan IEEE card :-(\n"
+"You may want report firmare revision (0x%X) and what the card support.\n"
+"I will try to make it work, but you should look for a better driver.\n", dev_info, firmware);
  			local->has_port3  = 1;
  			local->has_ibssid = 0;
  			local->has_mwo    = 0;
diff -X dontdiff -urp kern_oldest/drivers/net/wan/sbni.c kern_fix/drivers/net/wan/sbni.c
--- kern_oldest/drivers/net/wan/sbni.c	2006-05-01 11:55:40.000000000 -0700
+++ kern_fix/drivers/net/wan/sbni.c	2006-05-01 15:25:34.000000000 -0700
@@ -1552,13 +1552,13 @@ __setup( "sbni=", sbni_setup );
 static u32
 calc_crc32( u32  crc,  u8  *p,  u32  len )
 {
-	register u32  _crc __asm ( "ax" );
+	register u32  _crc;
 	_crc = crc;
 	
 	__asm __volatile (
 		"xorl	%%ebx, %%ebx\n"
-		"movl	%1, %%esi\n" 
-		"movl	%2, %%ecx\n" 
+		"movl	%2, %%esi\n" 
+		"movl	%3, %%ecx\n" 
 		"movl	$crc32tab, %%edi\n"
 		"shrl	$2, %%ecx\n"
 		"jz	1f\n"
@@ -1594,7 +1594,7 @@ calc_crc32( u32  crc,  u8  *p,  u32  len
 		"jnz	0b\n"
 
 	"1:\n"
-		"movl	%2, %%ecx\n"
+		"movl	%3, %%ecx\n"
 		"andl	$3, %%ecx\n"
 		"jz	2f\n"
 
@@ -1619,9 +1619,9 @@ calc_crc32( u32  crc,  u8  *p,  u32  len
 		"xorb	2(%%esi), %%bl\n"
 		"xorl	(%%edi,%%ebx,4), %%eax\n"
 	"2:\n"
-		:
-		: "a" (_crc), "g" (p), "g" (len)
-		: "ax", "bx", "cx", "dx", "si", "di"
+		: "=a" (_crc)
+		: "0" (_crc), "g" (p), "g" (len)
+		: "bx", "cx", "dx", "si", "di"
 	);
 
 	return  _crc;
diff -X dontdiff -urp kern_oldest/drivers/sensors/lm87.c kern_fix/drivers/sensors/lm87.c
--- kern_oldest/drivers/sensors/lm87.c	2006-05-01 11:55:44.000000000 -0700
+++ kern_fix/drivers/sensors/lm87.c	2006-05-01 15:25:34.000000000 -0700
@@ -1060,10 +1060,10 @@ MODULE_LICENSE("GPL");
 #endif
 
 MODULE_AUTHOR
-    ("Frodo Looijaard <frodol@dds.nl>,
-      Philip Edelbrock <phil@netroedge.com>, 
-      Mark Studebaker <mdsxyz123@yahoo.com>,
-      and Stephen Rousset <stephen.rousset@rocketlogix.com>");
+    ("Frodo Looijaard <frodol@dds.nl>, "
+      "Philip Edelbrock <phil@netroedge.com>, "
+      "Mark Studebaker <mdsxyz123@yahoo.com>, "
+      "and Stephen Rousset <stephen.rousset@rocketlogix.com>");
 
 MODULE_DESCRIPTION("LM87 driver");
 
