Index: linux-2.6.7/include/linux/dcache.h
===================================================================
--- linux-2.6.7.orig/include/linux/dcache.h	2005-03-23 23:28:49.669799416 +0800
+++ linux-2.6.7/include/linux/dcache.h	2005-03-23 23:38:25.648237384 +0800
@@ -86,6 +86,9 @@
 	spinlock_t d_lock;		/* per dentry lock */
 	struct inode *d_inode;		/* Where the name belongs to - NULL is
 					 * negative */
+        unsigned d_inum;                /* for cross-fs references (Lustre) */
+        unsigned d_mdsnum;              /* for cross-fs references (Lustre) */
+        unsigned d_generation;          /* for cross-fs references (Lustre) */
 	/*
 	 * The next three fields are touched by __d_lookup.  Place them here
 	 * so they all fit in a 16-byte range, with 16-byte alignment.
@@ -158,6 +161,8 @@
 #define DCACHE_UNHASHED		0x0010	
 #define DCACHE_LUSTRE_INVALID	0x0020	/* invalidated by Lustre */
 
+#define DCACHE_CROSS_REF       0x0040  /* entry points to inode on another MDS */
+
 extern spinlock_t dcache_lock;
 
 /**
