diff -rupN linux-2.6.7.old/include/linux/dcache.h linux-2.6.7/include/linux/dcache.h
--- linux-2.6.7.old/include/linux/dcache.h	2004-11-05 16:48:26.224926856 +0200
+++ linux-2.6.7/include/linux/dcache.h	2004-11-04 09:21:03.000000000 +0200
@@ -97,6 +97,7 @@ struct dentry {
         unsigned d_inum;                /* for cross-fs references (Lustre) */
         unsigned d_mdsnum;              /* for cross-fs references (Lustre) */
         unsigned d_generation;          /* for cross-fs references (Lustre) */
+        unsigned d_fid;                 /* for cross-fs references (Lustre) */
 	/*
 	 * The next three fields are touched by __d_lookup.  Place them here
 	 * so they all fit in a 16-byte range, with 16-byte alignment.
