Index: linux-2.6.5-7.276/include/linux/dcache.h
===================================================================
--- linux-2.6.5-7.276.orig/include/linux/dcache.h
+++ linux-2.6.5-7.276/include/linux/dcache.h
@@ -38,7 +38,6 @@ struct qstr {
 	const unsigned char * name;
 	unsigned int len;
 	unsigned int hash;
-	char name_str[0];
 };
 
 #include <linux/namei.h>
@@ -104,7 +103,6 @@ struct dentry {
  	struct rcu_head d_rcu;
 	struct dcookie_struct * d_cookie; /* cookie, if any */
 	unsigned long d_move_count;	/* to indicated moved dentry while lockless lookup */
-	struct qstr * d_qstr;		/* quick str ptr used in lockless lookup and concurrent d_move */
 	struct dentry * d_parent;	/* parent directory */
 	struct qstr d_name;
 	struct hlist_node d_hash;	/* lookup hash list */	
Index: linux-2.6.5-7.276/fs/dcache.c
===================================================================
--- linux-2.6.5-7.276.orig/fs/dcache.c
+++ linux-2.6.5-7.276/fs/dcache.c
@@ -775,7 +775,6 @@ struct dentry * d_alloc(struct dentry * 
 	dentry->d_parent = NULL;
 	dentry->d_move_count = 0;
 	dentry->d_sb = NULL;
-	dentry->d_qstr = &dentry->d_name;
 	dentry->d_name.name = str;
 	dentry->d_name.len = name->len;
 	dentry->d_name.hash = name->hash;
