--- linux-2.6.5.orig/drivers/block/ll_rw_blk.c	2005-03-16 11:08:47.404935605 -0800
+++ linux-2.6.5/drivers/block/ll_rw_blk.c	2005-03-16 10:57:38.197357003 -0800
@@ -2458,6 +2458,8 @@ static inline void blk_partition_remap(s
 	}
 }
 
+int dev_check_rdonly(struct block_device *bdev);
+
 /**
  * generic_make_request: hand a buffer to its device driver for I/O
  * @bio:  The bio describing the location in memory and on the device.
@@ -2546,6 +2548,13 @@ end_io:
 		if (test_bit(QUEUE_FLAG_DEAD, &q->queue_flags))
 			goto end_io;
 
+		/* this is cfs's dev_rdonly check */
+		if (bio->bi_rw == WRITE &&
+		                dev_check_rdonly(bio->bi_bdev)) {
+		        bio_endio(bio, bio->bi_size, 0);
+		        break;
+		}
+
 		/*
 		 * If this device has partitions, remap block n
 		 * of partition p to block n+start(p) of the disk.
@@ -3078,6 +3087,91 @@ void swap_io_context(struct io_context *
 }
 
 /*
+ * Debug code for turning block devices "read-only" (will discard writes
+ * silently).  This is for filesystem crash/recovery testing.
+ */
+struct deventry {
+        dev_t dev;
+        struct deventry *next;
+};
+
+static struct deventry *devlist = NULL;
+static spinlock_t devlock = SPIN_LOCK_UNLOCKED; 
+
+int dev_check_rdonly(struct block_device *bdev) 
+{
+        struct deventry *cur;
+        if (!bdev) return 0;
+        spin_lock(&devlock);
+        cur = devlist;
+        while(cur) {
+                if (bdev->bd_dev == cur->dev) {
+                        spin_unlock(&devlock);
+                        return 1;
+                }
+                cur = cur->next;
+        }
+        spin_unlock(&devlock);
+        return 0;
+}
+
+void dev_set_rdonly(struct block_device *bdev)
+{
+        struct deventry *newdev, *cur;
+
+        if (!bdev) 
+	        return;
+        newdev = kmalloc(sizeof(struct deventry), GFP_KERNEL);
+        if (!newdev) 
+	        return;
+
+        spin_lock(&devlock);
+        cur = devlist;
+        while(cur) {
+                if (bdev->bd_dev == cur->dev) {
+                        spin_unlock(&devlock);
+                        kfree(newdev);
+                        return;
+                }
+                cur = cur->next;
+        }
+        newdev->dev = bdev->bd_dev;
+        newdev->next = devlist;
+        devlist = newdev;
+        spin_unlock(&devlock);
+        printk(KERN_WARNING "Turning device %s read-only\n",
+               bdev->bd_disk ? bdev->bd_disk->disk_name : "?");
+}
+
+void dev_clear_rdonly(struct block_device *bdev) 
+{
+        struct deventry *cur, *last = NULL;
+        if (!bdev) return;
+        spin_lock(&devlock);
+        cur = devlist;
+        while(cur) {
+                if (bdev->bd_dev == cur->dev) {
+                        if (last) 
+                                last->next = cur->next;
+                        else
+                                devlist = cur->next;
+                        spin_unlock(&devlock);
+                        kfree(cur);
+                        printk(KERN_WARNING "Removing read-only on %s\n",
+			       bdev->bd_disk ? bdev->bd_disk->disk_name : "?");
+			return;
+                }
+                last = cur;
+                cur = cur->next;
+        }
+        spin_unlock(&devlock);
+}
+
+EXPORT_SYMBOL(dev_set_rdonly);
+EXPORT_SYMBOL(dev_clear_rdonly);
+EXPORT_SYMBOL(dev_check_rdonly);
+
+/*
  * sysfs parts below
  */
 struct queue_sysfs_entry {
