Index: lum/mm/filemap.c
===================================================================
--- lum.orig/mm/filemap.c	2004-04-25 14:58:10.000000000 -0400
+++ lum/mm/filemap.c	2004-04-25 16:23:32.000000000 -0400
@@ -1614,8 +1614,8 @@
 		new_iobuf = 1;
 	}
 
-	blocksize = 1 << inode->i_blkbits;
-	blocksize_bits = inode->i_blkbits;
+	blocksize = 512 /*1 << inode->i_blkbits*/;
+	blocksize_bits = 9 /*inode->i_blkbits*/;
 	blocksize_mask = blocksize - 1;
 	chunk_size = KIO_MAX_ATOMIC_IO << 10;
 
