Index: linux-2.6.10/fs/dcache.c
===================================================================
--- linux-2.6.10.orig/fs/dcache.c	2006-10-21 11:52:54.000000000 +0800
+++ linux-2.6.10/fs/dcache.c	2006-10-21 13:04:55.000000000 +0800
@@ -1664,6 +1664,7 @@
 
 extern void bdev_cache_init(void);
 extern void chrdev_init(void);
+extern void dynlock_cache_init(void);
 
 void __init vfs_caches_init_early(void)
 {
@@ -1693,6 +1694,7 @@
 	mnt_init(mempages);
 	bdev_cache_init();
 	chrdev_init();
+	dynlock_cache_init();
 }
 
 EXPORT_SYMBOL(d_alloc);
Index: linux-2.6.10/lib/Makefile
===================================================================
--- linux-2.6.10.orig/lib/Makefile	2004-12-25 05:33:50.000000000 +0800
+++ linux-2.6.10/lib/Makefile	2006-10-21 13:08:20.000000000 +0800
@@ -5,7 +5,7 @@
 lib-y := errno.o ctype.o string.o vsprintf.o cmdline.o \
 	 bust_spinlocks.o rbtree.o radix-tree.o dump_stack.o \
 	 kobject.o kref.o idr.o div64.o parser.o int_sqrt.o \
-	 bitmap.o extable.o kobject_uevent.o
+	 bitmap.o extable.o kobject_uevent.o dynlocks.o
 
 ifeq ($(CONFIG_DEBUG_KOBJECT),y)
 CFLAGS_kobject.o += -DDEBUG
