Index: linux-2.6.16.46-0.14/fs/dcache.c
===================================================================
--- linux-2.6.16.46-0.14.orig/fs/dcache.c	2007-08-30 05:43:46.000000000 +0300
+++ linux-2.6.16.46-0.14/fs/dcache.c	2007-08-30 05:45:51.000000000 +0300
@@ -1849,6 +1849,7 @@
 
 extern void bdev_cache_init(void);
 extern void chrdev_init(void);
+extern void dynlock_cache_init(void);
 
 void __init vfs_caches_init_early(void)
 {
@@ -1879,6 +1880,7 @@
 	mnt_init(mempages);
 	bdev_cache_init();
 	chrdev_init();
+	dynlock_cache_init();
 }
 
 EXPORT_SYMBOL(d_alloc);
Index: linux-2.6.16.46-0.14/lib/Makefile
===================================================================
--- linux-2.6.16.46-0.14.orig/lib/Makefile	2007-07-18 08:02:26.000000000 +0300
+++ linux-2.6.16.46-0.14/lib/Makefile	2007-08-30 05:47:17.000000000 +0300
@@ -5,7 +5,7 @@
 lib-y := errno.o ctype.o string.o vsprintf.o cmdline.o \
 	 bust_spinlocks.o rbtree.o radix-tree.o dump_stack.o \
 	 idr.o div64.o int_sqrt.o bitmap.o extable.o prio_tree.o \
-	 sha1.o
+	 sha1.o dynlocks.o
 
 lib-y	+= kobject.o kref.o kobject_uevent.o klist.o
 
