Index: linux/fs/dcache.c
===================================================================
--- linux.orig/fs/dcache.c
+++ linux/fs/dcache.c
@@ -1678,6 +1678,7 @@ EXPORT_SYMBOL(d_genocide);
 
 extern void bdev_cache_init(void);
 extern void chrdev_init(void);
+extern void dynlock_cache_init(void);
 
 void __init vfs_caches_init_early(void)
 {
@@ -1707,6 +1708,7 @@ void __init vfs_caches_init(unsigned lon
 	mnt_init(mempages);
 	bdev_cache_init();
 	chrdev_init();
+	dynlock_cache_init();
 }
 
 EXPORT_SYMBOL(d_alloc);
Index: linux/lib/Makefile
===================================================================
--- linux.orig/lib/Makefile
+++ linux/lib/Makefile
@@ -6,7 +6,7 @@
 lib-y := errno.o ctype.o string.o vsprintf.o cmdline.o \
 	 bust_spinlocks.o rbtree.o radix-tree.o dump_stack.o \
 	 kobject.o kref.o idr.o div64.o parser.o int_sqrt.o \
-	 bitmap.o extable.o
+	 bitmap.o extable.o dynlocks.o
 
 obj-y := sort.o
 
