diff -p -u linux-2.4/drivers/net/e1000/e1000_main.c /tmp/drivers/net/e1000/e1000_main.c 
--- linux-2.4/drivers/net/e1000/e1000_main.c	2004-06-04 17:07:49.000000000 +0100
+++ /tmp/drivers/net/e1000/e1000_main.c	2004-06-04 17:07:42.000000000 +0100
@@ -462,6 +462,7 @@ e1000_probe(struct pci_dev *pdev,
 	netdev->mem_start = mmio_start;
 	netdev->mem_end = mmio_start + mmio_len;
 	netdev->base_addr = adapter->hw.io_base;
+	netdev->irq = pdev->irq;
 
 	adapter->bd_number = cards_found;
 

