Index: linux-stage/fs/ext3/super.c
===================================================================
--- linux-stage.orig/fs/ext3/super.c	2004-04-02 16:40:18.000000000 -0500
+++ linux-stage/fs/ext3/super.c	2004-04-02 16:40:18.000000000 -0500
@@ -115,6 +115,8 @@
 		handle->h_err = err;
 }
 
+EXPORT_SYMBOL(ext3_journal_abort_handle);
+
 static char error_buf[1024];
 
 /* Deal with the reporting of failure conditions on a filesystem such as
@@ -1774,6 +1776,8 @@
 	return ret;
 }
 
+EXPORT_SYMBOL(ext3_force_commit);
+
 /*
  * Ext3 always journals updates to the superblock itself, so we don't
  * have to propagate any other updates to the superblock on disk at this
@@ -2088,6 +2092,10 @@
                         unsigned long *blocks, int *created, int create);
 EXPORT_SYMBOL(ext3_map_inode_page);
 
+EXPORT_SYMBOL(ext3_xattr_get);
+EXPORT_SYMBOL(ext3_xattr_set_handle);
+EXPORT_SYMBOL(ext3_bread);
+
 MODULE_AUTHOR("Remy Card, Stephen Tweedie, Andrew Morton, Andreas Dilger, Theodore Ts'o and others");
 MODULE_DESCRIPTION("Second Extended Filesystem with journaling extensions");
 MODULE_LICENSE("GPL");
