 mm/filemap.c |    2 ++
 1 files changed, 2 insertions(+)

--- linux-2.4.20-l19/mm/filemap.c~export-add_to_page_cache	2003-07-15 03:05:47.000000000 -0500
+++ linux-2.4.20-l19-zab/mm/filemap.c	2003-07-15 15:33:01.000000000 -0500
@@ -670,6 +670,7 @@ void add_to_page_cache(struct page * pag
 	spin_unlock(&pagecache_lock);
 	lru_cache_add(page);
 }
+EXPORT_SYMBOL_GPL(add_to_page_cache);
 
 int add_to_page_cache_unique(struct page * page,
 	struct address_space *mapping, unsigned long offset,
@@ -692,6 +693,7 @@ int add_to_page_cache_unique(struct page
 		lru_cache_add(page);
 	return err;
 }
+EXPORT_SYMBOL_GPL(add_to_page_cache_unique);
 
 /*
  * This adds the requested page to the page cache if it isn't already there,

_
