Index: linux-2.6.16.i686/kernel/sched.c
===================================================================
--- linux-2.6.16.i686.orig/kernel/sched.c	2006-05-30 15:47:15.000000000 +0800
+++ linux-2.6.16.i686/kernel/sched.c	2006-05-30 21:21:07.000000000 +0800
@@ -4240,7 +4240,7 @@
 	return list_entry(p->sibling.next,struct task_struct,sibling);
 }
 
-static void show_task(task_t *p)
+void show_task(task_t *p)
 {
 	task_t *relative;
 	unsigned state;
@@ -4290,9 +4290,10 @@
 	else
 		printk(" (NOTLB)\n");
 
-	if (state != TASK_RUNNING)
+	if (state != TASK_RUNNING || p == current)
 		show_stack(p, NULL);
 }
+EXPORT_SYMBOL(show_task);
 
 void show_state(void)
 {
