Index: linux-2.6.18-128.1.6/mm/page_alloc.c
===================================================================
--- linux-2.6.18-128.1.6.orig/mm/page_alloc.c	2009-04-14 21:05:42.000000000 -0600
+++ linux-2.6.18-128.1.6/mm/page_alloc.c	2009-06-02 23:24:34.000000000 -0600
@@ -877,6 +877,7 @@
 	}
 	return 1;
 }
+EXPORT_SYMBOL(nr_free_buffer_pages);
 
 /*
  * get_page_from_freeliest goes through the zonelist trying to allocate
