Index: linux-2.6/fs/filesystems.c
===================================================================
--- linux-2.6.orig/fs/filesystems.c	2006-07-15 16:08:35.000000000 +0800
+++ linux-2.6/fs/filesystems.c	2006-07-15 16:14:19.000000000 +0800
@@ -29,7 +29,9 @@
  */
 
 static struct file_system_type *file_systems;
-static DEFINE_RWLOCK(file_systems_lock);
+DEFINE_RWLOCK(file_systems_lock);
+
+EXPORT_SYMBOL(file_systems_lock);
 
 /* WARNING: This can be used only if we _already_ own a reference */
 void get_filesystem(struct file_system_type *fs)
Index: linux-2.6/include/linux/fs.h
===================================================================
--- linux-2.6.orig/include/linux/fs.h	2006-07-15 16:10:37.000000000 +0800
+++ linux-2.6/include/linux/fs.h	2006-07-15 16:14:19.000000000 +0800
@@ -1768,6 +1768,7 @@ static inline ssize_t blockdev_direct_IO
 
 extern const struct file_operations generic_ro_fops;
 
+extern rwlock_t file_systems_lock;
 #define special_file(m) (S_ISCHR(m)||S_ISBLK(m)||S_ISFIFO(m)||S_ISSOCK(m))
 
 extern int vfs_readlink(struct dentry *, char __user *, int, const char *);
Index: linux-2.6/fs/namespace.c
===================================================================
--- linux-2.6.orig/fs/namespace.c	2006-07-15 16:10:33.000000000 +0800
+++ linux-2.6/fs/namespace.c	2006-07-15 16:14:19.000000000 +0800
@@ -1641,6 +1641,7 @@ void set_fs_pwd(struct fs_struct *fs, st
 		mntput(old_pwdmnt);
 	}
 }
+EXPORT_SYMBOL(set_fs_pwd);
 
 static void chroot_fs_refs(struct nameidata *old_nd, struct nameidata *new_nd)
 {
Index: linux-2.6/kernel/exit.c
===================================================================
--- linux-2.6.orig/kernel/exit.c	2006-07-15 16:08:34.000000000 +0800
+++ linux-2.6/kernel/exit.c	2006-07-15 16:14:19.000000000 +0800
@@ -305,6 +305,8 @@ static void reparent_to_init(void)
 	switch_uid(INIT_USER);
 }
 
+EXPORT_SYMBOL(reparent_to_init);
+
 void __set_special_pids(pid_t session, pid_t pgrp)
 {
 	struct task_struct *curr = current->group_leader;
Index: linux-2.6/fs/dcache.c
===================================================================
--- linux-2.6.orig/fs/dcache.c	2006-07-15 16:14:00.000000000 +0800
+++ linux-2.6/fs/dcache.c	2006-07-15 16:14:19.000000000 +0800
@@ -1628,6 +1628,7 @@ int is_subdir(struct dentry * new_dentry
 
 	return result;
 }
+EXPORT_SYMBOL(is_subdir);
 
 void d_genocide(struct dentry *root)
 {
