Index: linux-2.6.9-5.0.3.EL/arch/ia64/kernel/smpboot.c
===================================================================
--- linux-2.6.9-5.0.3.EL.orig/arch/ia64/kernel/smpboot.c	2004-10-18 15:53:43.000000000 -0600
+++ linux-2.6.9-5.0.3.EL/arch/ia64/kernel/smpboot.c	2005-09-28 16:02:16.000000000 -0600
@@ -485,6 +485,7 @@
 EXPORT_SYMBOL(cpu_to_node_map);
 /* which logical CPUs are on which nodes */
 cpumask_t node_to_cpu_mask[MAX_NUMNODES] __cacheline_aligned;
+EXPORT_SYMBOL(node_to_cpu_mask);
 
 /*
  * Build cpu to node mapping and initialize the per node cpu masks.
Index: linux-2.6.9-5.0.3.EL/arch/i386/kernel/smpboot.c
===================================================================
--- linux-2.6.9-5.0.3.EL.orig/arch/i386/kernel/smpboot.c	2004-10-18 15:54:08.000000000 -0600
+++ linux-2.6.9-5.0.3.EL/arch/i386/kernel/smpboot.c	2005-09-28 16:28:12.000000000 -0600
@@ -474,6 +474,7 @@
 /* which logical CPUs are on which nodes */
 cpumask_t node_2_cpu_mask[MAX_NUMNODES] =
 				{ [0 ... MAX_NUMNODES-1] = CPU_MASK_NONE };
+EXPORT_SYMBOL(node_2_cpu_mask);
 /* which node each logical CPU is on */
 int cpu_2_node[NR_CPUS] = { [0 ... NR_CPUS-1] = 0 };
 EXPORT_SYMBOL(cpu_2_node);
