Index: linux-2.6.22.5/fs/filesystems.c
===================================================================
--- linux-2.6.22.5.orig/fs/filesystems.c	2007-08-22 17:23:54.000000000 -0600
+++ linux-2.6.22.5/fs/filesystems.c	2008-02-21 00:58:18.000000000 -0700
@@ -28,7 +28,9 @@
  */
 
 static struct file_system_type *file_systems;
-static DEFINE_RWLOCK(file_systems_lock);
+DEFINE_RWLOCK(file_systems_lock);
+
+EXPORT_SYMBOL(file_systems_lock);
 
 /* WARNING: This can be used only if we _already_ own a reference */
 void get_filesystem(struct file_system_type *fs)
Index: linux-2.6.22.5/fs/namespace.c
===================================================================
--- linux-2.6.22.5.orig/fs/namespace.c	2007-08-22 17:23:54.000000000 -0600
+++ linux-2.6.22.5/fs/namespace.c	2008-02-21 00:58:18.000000000 -0700
@@ -1617,6 +1617,7 @@
 		mntput(old_pwdmnt);
 	}
 }
+EXPORT_SYMBOL(set_fs_pwd);
 
 static void chroot_fs_refs(struct nameidata *old_nd, struct nameidata *new_nd)
 {
Index: linux-2.6.22.5/fs/dcache.c
===================================================================
--- linux-2.6.22.5.orig/fs/dcache.c	2008-02-21 00:57:02.000000000 -0700
+++ linux-2.6.22.5/fs/dcache.c	2008-02-21 00:58:18.000000000 -0700
@@ -2027,6 +2027,7 @@
 
 	return result;
 }
+EXPORT_SYMBOL(is_subdir);
 
 void d_genocide(struct dentry *root)
 {
Index: linux-2.6.22.5/include/linux/fs.h
===================================================================
--- linux-2.6.22.5.orig/include/linux/fs.h	2008-02-21 00:56:39.000000000 -0700
+++ linux-2.6.22.5/include/linux/fs.h	2008-02-21 00:58:18.000000000 -0700
@@ -1861,6 +1861,7 @@
 
 extern const struct file_operations generic_ro_fops;
 
+extern rwlock_t file_systems_lock;
 #define special_file(m) (S_ISCHR(m)||S_ISBLK(m)||S_ISFIFO(m)||S_ISSOCK(m))
 
 extern int vfs_readlink(struct dentry *, char __user *, int, const char *);
