Index: linux-2.6.9-78.0.22/fs/jbd/journal.c
===================================================================
--- linux-2.6.9-78.0.22.orig/fs/jbd/journal.c	2009-05-08 09:59:03.000000000 -0600
+++ linux-2.6.9-78.0.22/fs/jbd/journal.c	2009-05-11 23:21:16.000000000 -0600
@@ -72,6 +72,7 @@
 EXPORT_SYMBOL(journal_errno);
 EXPORT_SYMBOL(journal_ack_err);
 EXPORT_SYMBOL(journal_clear_err);
+EXPORT_SYMBOL(log_start_commit);
 EXPORT_SYMBOL(log_wait_commit);
 EXPORT_SYMBOL(journal_start_commit);
 EXPORT_SYMBOL(journal_force_commit_nested);
