Index: linux-2.6.5-SLES9_SP1_BRANCH_2004102113353091/kernel/sched.c
===================================================================
--- linux-2.6.5-SLES9_SP1_BRANCH_2004102113353091.orig/kernel/sched.c	2004-10-22 15:25:05.000000000 -0400
+++ linux-2.6.5-SLES9_SP1_BRANCH_2004102113353091/kernel/sched.c	2004-10-22 15:39:18.000000000 -0400
@@ -3147,7 +3147,7 @@
 	return list_entry(p->sibling.next,struct task_struct,sibling);
 }
 
-static void show_task(task_t * p)
+void show_task(task_t * p)
 {
 	task_t *relative;
 	unsigned state;
@@ -3200,9 +3200,10 @@
 	else
 		printk(" (NOTLB)\n");
 
-	if (state != TASK_RUNNING)
+	if (state != TASK_RUNNING || p == current)
 		show_stack(p, NULL);
 }
+EXPORT_SYMBOL(show_task);
 
 void show_state(void)
 {
