Index: kernel-l0405/kernel/sched.c
===================================================================
--- kernel-l0405.orig/kernel/sched.c	2003-11-06 16:15:20.000000000 -0800
+++ kernel-l0405/kernel/sched.c	2005-04-05 14:44:27.000000000 -0700
@@ -1627,7 +1627,7 @@
 	return retval;
 }
 
-static void show_task(task_t * p)
+void show_task(task_t * p)
 {
 	unsigned long free = 0;
 	int state;
Index: kernel-l0405/kernel/ksyms.c
===================================================================
--- kernel-l0405.orig/kernel/ksyms.c	2005-04-05 14:44:15.000000000 -0700
+++ kernel-l0405/kernel/ksyms.c	2005-04-05 14:44:50.000000000 -0700
@@ -55,6 +55,7 @@
 #include <linux/unistd.h>
 #include <linux/bdf_prm.h>
 
+extern void show_task(task_t *);
 
 #if defined(CONFIG_PROC_FS)
 #include <linux/proc_fs.h>
@@ -684,6 +685,7 @@
 
 /* debug */
 EXPORT_SYMBOL(dump_stack);
+EXPORT_SYMBOL(show_task);
 
 #if defined(CONFIG_KDB_USB)
 #include <linux/kdb.h>
