Index: linux-2.6/fs/super.c
===================================================================
--- linux-2.6.orig/fs/super.c	2006-07-20 10:51:39.000000000 +0800
+++ linux-2.6/fs/super.c	2006-07-20 10:51:59.000000000 +0800
@@ -877,6 +877,8 @@ do_kern_mount(const char *fstype, int fl
 	return mnt;
 }
 
+EXPORT_SYMBOL_GPL(do_kern_mount);
+
 struct vfsmount *kern_mount(struct file_system_type *type)
 {
 	return vfs_kern_mount(type, 0, type->name, NULL);
