Index: linux-bgl/kernel/sched.c
===================================================================
--- linux-bgl.orig/kernel/sched.c	2003-07-02 08:43:33.000000000 -0700
+++ linux-bgl/kernel/sched.c	2004-10-26 23:37:44.314193755 -0700
@@ -1124,7 +1124,7 @@
 	return retval;
 }
 
-static void show_task(struct task_struct * p)
+void show_task(struct task_struct * p)
 {
 	unsigned long free = 0;
 	int state;
Index: linux-bgl/kernel/ksyms.c
===================================================================
--- linux-bgl.orig/kernel/ksyms.c	2004-10-26 23:23:00.518654978 -0700
+++ linux-bgl/kernel/ksyms.c	2004-10-26 23:38:29.289071295 -0700
@@ -76,6 +76,7 @@
 };
 #endif
 
+void show_task(struct task_struct *);
 
 EXPORT_SYMBOL(inter_module_register);
 EXPORT_SYMBOL(inter_module_unregister);
@@ -595,3 +596,6 @@
 
 EXPORT_SYMBOL(tasklist_lock);
 EXPORT_SYMBOL(pidhash);
+
+/* debug */
+EXPORT_SYMBOL(show_task);
