Index: linux-2.6.10/fs/ext3/super.c
===================================================================
--- linux-2.6.10.orig/fs/ext3/super.c	2005-03-31 18:44:38.935933960 +0800
+++ linux-2.6.10/fs/ext3/super.c	2005-03-31 18:46:03.008153040 +0800
@@ -123,6 +123,8 @@
 	journal_abort_handle(handle);
 }
 
+EXPORT_SYMBOL(ext3_journal_abort_handle);
+ 
 /* Deal with the reporting of failure conditions on a filesystem such as
  * inconsistencies detected or read IO failures.
  *
@@ -2016,6 +2018,8 @@
 	return ret;
 }
 
+EXPORT_SYMBOL(ext3_force_commit);
+
 /*
  * Ext3 always journals updates to the superblock itself, so we don't
  * have to propagate any other updates to the superblock on disk at this
@@ -2447,6 +2451,10 @@
                         unsigned long *blocks, int *created, int create);
 EXPORT_SYMBOL(ext3_map_inode_page);
 
+EXPORT_SYMBOL(ext3_xattr_get);
+EXPORT_SYMBOL(ext3_xattr_set_handle);
+EXPORT_SYMBOL(ext3_bread);
+
 MODULE_AUTHOR("Remy Card, Stephen Tweedie, Andrew Morton, Andreas Dilger, Theodore Ts'o and others");
 MODULE_DESCRIPTION("Second Extended Filesystem with journaling extensions");
 MODULE_LICENSE("GPL");
