Index: linux-2.4.21/kernel/ksyms.c
===================================================================
--- linux-2.4.21.orig/kernel/ksyms.c	2004-11-01 22:00:03.000000000 -0500
+++ linux-2.4.21/kernel/ksyms.c	2004-11-01 22:07:07.000000000 -0500
@@ -86,6 +86,7 @@
 };
 #endif
 
+extern void show_task(task_t *);
 
 EXPORT_SYMBOL(inter_module_register);
 EXPORT_SYMBOL(inter_module_unregister);
@@ -670,6 +671,7 @@
 extern void check_tasklist_locked(void);
 EXPORT_SYMBOL_GPL(check_tasklist_locked);
 EXPORT_SYMBOL(dump_stack);
+EXPORT_SYMBOL(show_task);
 
 EXPORT_SYMBOL_GPL(netdump_func);
 EXPORT_SYMBOL_GPL(netdump_mode);
