Index: linux-2.6.16.i686/arch/i386/kernel/smpboot.c
===================================================================
--- linux-2.6.16.i686.orig/arch/i386/kernel/smpboot.c	2006-05-30 15:47:03.000000000 +0800
+++ linux-2.6.16.i686/arch/i386/kernel/smpboot.c	2006-05-30 21:22:02.000000000 +0800
@@ -579,6 +579,7 @@
 /* which logical CPUs are on which nodes */
 cpumask_t node_2_cpu_mask[MAX_NUMNODES] __read_mostly =
 				{ [0 ... MAX_NUMNODES-1] = CPU_MASK_NONE };
+EXPORT_SYMBOL(node_2_cpu_mask);
 /* which node each logical CPU is on */
 int cpu_2_node[NR_CPUS] __read_mostly = { [0 ... NR_CPUS-1] = 0 };
 EXPORT_SYMBOL(cpu_2_node);
