Index: linux-2.4.24/kernel/ksyms.c
===================================================================
--- linux-2.4.24.orig/kernel/ksyms.c	2004-10-22 14:00:00.000000000 -0400
+++ linux-2.4.24/kernel/ksyms.c	2004-10-22 14:03:54.000000000 -0400
@@ -70,6 +70,8 @@
 extern spinlock_t dma_spin_lock;
 extern int panic_timeout;
 
+extern void show_task(struct task_struct *);
+
 #ifdef CONFIG_MODVERSIONS
 const struct module_symbol __export_Using_Versions
 __attribute__((section("__ksymtab"))) = {
@@ -619,6 +621,7 @@
 
 /* debug */
 EXPORT_SYMBOL(dump_stack);
+EXPORT_SYMBOL(show_task);
 
 /* To match ksyms with System.map */
 extern const char _end[];
Index: linux-2.4.24/kernel/sched.c
===================================================================
--- linux-2.4.24.orig/kernel/sched.c	2003-11-28 13:26:21.000000000 -0500
+++ linux-2.4.24/kernel/sched.c	2004-10-22 14:10:10.000000000 -0400
@@ -1177,7 +1177,7 @@
 	return retval;
 }
 
-static void show_task(struct task_struct * p)
+void show_task(struct task_struct * p)
 {
 	unsigned long free = 0;
 	int state;
