


 0 files changed

--- linux-2.4.18-17.8.0/fs/ext3/Makefile~exports	2002-12-06 14:52:29.000000000 -0800
+++ linux-2.4.18-17.8.0-zab/fs/ext3/Makefile	2002-12-06 14:52:29.000000000 -0800
@@ -9,6 +9,8 @@
 
 O_TARGET := ext3.o
 
+export-objs :=	super.o
+
 obj-y    := balloc.o bitmap.o dir.o file.o fsync.o ialloc.o inode.o \
 		ioctl.o namei.o super.o symlink.o
 obj-m    := $(O_TARGET)
--- linux-2.4.18-17.8.0/fs/ext3/super.c~exports	2002-12-06 14:52:29.000000000 -0800
+++ linux-2.4.18-17.8.0-zab/fs/ext3/super.c	2002-12-06 14:52:29.000000000 -0800
@@ -1746,7 +1746,7 @@ static void __exit exit_ext3_fs(void)
 	unregister_filesystem(&ext3_fs_type);
 }
 
-EXPORT_NO_SYMBOLS;
+EXPORT_SYMBOL(ext3_bread);
 
 MODULE_AUTHOR("Remy Card, Stephen Tweedie, Andrew Morton, Andreas Dilger, Theodore Ts'o and others");
 MODULE_DESCRIPTION("Second Extended Filesystem with journaling extensions");
--- linux-2.4.18-17.8.0/include/linux/fs.h~exports	2002-12-06 14:52:29.000000000 -0800
+++ linux-2.4.18-17.8.0-zab/include/linux/fs.h	2002-12-06 14:52:29.000000000 -0800
@@ -1046,6 +1046,7 @@ extern int unregister_filesystem(struct 
 extern struct vfsmount *kern_mount(struct file_system_type *);
 extern int may_umount(struct vfsmount *);
 extern long do_mount(char *, char *, char *, unsigned long, void *);
+struct vfsmount *do_kern_mount(const char *type, int flags, char *name, void *data);
 extern void umount_tree(struct vfsmount *);
 
 #define kern_umount mntput
--- linux-2.4.18-17.8.0/kernel/ksyms.c~exports	2002-12-06 14:52:29.000000000 -0800
+++ linux-2.4.18-17.8.0-zab/kernel/ksyms.c	2002-12-06 14:52:29.000000000 -0800
@@ -306,6 +306,11 @@ EXPORT_SYMBOL_GPL(buffermem_pages);
 EXPORT_SYMBOL_GPL(nr_free_pages);
 EXPORT_SYMBOL_GPL(page_cache_size);
 
+/* lustre */
+EXPORT_SYMBOL(panic_notifier_list);
+EXPORT_SYMBOL(pagecache_lock_cacheline);
+EXPORT_SYMBOL(do_kern_mount);
+
 /* for stackable file systems (lofs, wrapfs, cryptfs, etc.) */
 EXPORT_SYMBOL(default_llseek);
 EXPORT_SYMBOL(dentry_open);

_
