Index: linux-2.6.0-test6/fs/filesystems.c
===================================================================
--- linux-2.6.0-test6.orig/fs/filesystems.c	2003-09-28 08:50:40.000000000 +0800
+++ linux-2.6.0-test6/fs/filesystems.c	2003-10-07 18:07:58.000000000 +0800
@@ -27,7 +27,7 @@
  */
 
 static struct file_system_type *file_systems;
-static rwlock_t file_systems_lock = RW_LOCK_UNLOCKED;
+rwlock_t file_systems_lock = RW_LOCK_UNLOCKED;
 
 /* WARNING: This can be used only if we _already_ own a reference */
 void get_filesystem(struct file_system_type *fs)
Index: linux-2.6.0-test6/include/linux/fs.h
===================================================================
--- linux-2.6.0-test6.orig/include/linux/fs.h	2003-10-07 16:09:11.000000000 +0800
+++ linux-2.6.0-test6/include/linux/fs.h	2003-10-09 19:47:24.996871688 +0800
@@ -1385,6 +1385,7 @@
 
 extern struct file_operations generic_ro_fops;
 
+extern rwlock_t file_systems_lock;
 #define special_file(m) (S_ISCHR(m)||S_ISBLK(m)||S_ISFIFO(m)||S_ISSOCK(m))
 
 extern int vfs_readlink(struct dentry *, char __user *, int, const char *);
Index: linux-2.6.0-test6/kernel/ksyms.c
===================================================================
--- linux-2.6.0-test6.orig/kernel/ksyms.c	2003-10-07 16:09:09.000000000 +0800
+++ linux-2.6.0-test6/kernel/ksyms.c	2003-10-07 18:07:58.000000000 +0800
@@ -393,3 +393,4 @@
 EXPORT_SYMBOL(dump_oncpu);
 EXPORT_SYMBOL(dump_function_ptr);
 #endif
+EXPORT_SYMBOL(file_systems_lock);
Index: linux-2.6.0-test6/include/linux/ext2_fs_sb.h
===================================================================
--- linux-2.6.0-test6.orig/include/linux/ext2_fs_sb.h	2003-09-28 08:50:57.000000000 +0800
+++ linux-2.6.0-test6/include/linux/ext2_fs_sb.h	2003-10-07 18:07:58.000000000 +0800
@@ -16,9 +16,11 @@
 #ifndef _LINUX_EXT2_FS_SB
 #define _LINUX_EXT2_FS_SB
 
+#ifndef EXT_INCLUDE
+#define EXT_INCLUDE
 #include <linux/blockgroup_lock.h>
 #include <linux/percpu_counter.h>
-
+#endif
 /*
  * second extended-fs super-block data in memory
  */
Index: linux-2.6.0-test6/include/linux/ext3_fs_sb.h
===================================================================
--- linux-2.6.0-test6.orig/include/linux/ext3_fs_sb.h	2003-09-28 08:51:00.000000000 +0800
+++ linux-2.6.0-test6/include/linux/ext3_fs_sb.h	2003-10-07 18:07:58.000000000 +0800
@@ -19,9 +19,12 @@
 #ifdef __KERNEL__
 #include <linux/timer.h>
 #include <linux/wait.h>
+#ifndef EXT_INCLUDE
+#define EXT_INCLUDE
 #include <linux/blockgroup_lock.h>
 #include <linux/percpu_counter.h>
 #endif
+#endif
 
 /*
  * third extended-fs super-block data in memory
