Index: linux-2.6.18.8/arch/i386/kernel/smpboot.c
===================================================================
--- linux-2.6.18.8.orig/arch/i386/kernel/smpboot.c	2007-06-05 13:20:25.000000000 +0200
+++ linux-2.6.18.8/arch/i386/kernel/smpboot.c	2007-06-05 13:20:33.000000000 +0200
@@ -607,6 +607,7 @@ extern struct {
 /* which logical CPUs are on which nodes */
 cpumask_t node_2_cpu_mask[MAX_NUMNODES] __read_mostly =
 				{ [0 ... MAX_NUMNODES-1] = CPU_MASK_NONE };
+EXPORT_SYMBOL(node_2_cpu_mask);
 /* which node each logical CPU is on */
 int cpu_2_node[NR_CPUS] __read_mostly = { [0 ... NR_CPUS-1] = 0 };
 EXPORT_SYMBOL(cpu_2_node);
Index: linux-2.6.18.8/arch/ia64/kernel/numa.c
===================================================================
--- linux-2.6.18.8.orig/arch/ia64/kernel/numa.c	2007-06-05 13:21:04.000000000 +0200
+++ linux-2.6.18.8/arch/ia64/kernel/numa.c	2007-06-05 13:21:28.000000000 +0200
@@ -28,6 +28,7 @@ u16 cpu_to_node_map[NR_CPUS] __cacheline
 EXPORT_SYMBOL(cpu_to_node_map);
 
 cpumask_t node_to_cpu_mask[MAX_NUMNODES] __cacheline_aligned;
+EXPORT_SYMBOL(node_to_cpu_mask);
 
 void __cpuinit map_cpu_to_node(int cpu, int nid)
 {
