


Index: linux-ion/kernel/ksyms.c
===================================================================
--- linux-ion.orig/kernel/ksyms.c	2004-07-28 14:34:57.000000000 -0700
+++ linux-ion/kernel/ksyms.c	2004-09-27 15:04:52.000000000 -0700
@@ -286,6 +286,10 @@
 EXPORT_SYMBOL(dcache_readdir);
 EXPORT_SYMBOL(dcache_dir_ops);
 
+/* lustre */
+EXPORT_SYMBOL(panic_notifier_list);
+EXPORT_SYMBOL(do_kern_mount);
+
 /* for stackable file systems (lofs, wrapfs, cryptfs, etc.) */
 EXPORT_SYMBOL(default_llseek);
 EXPORT_SYMBOL(dentry_open);
Index: linux-ion/include/linux/fs.h
===================================================================
--- linux-ion.orig/include/linux/fs.h	2004-07-28 14:34:57.000000000 -0700
+++ linux-ion/include/linux/fs.h	2004-09-27 15:04:52.000000000 -0700
@@ -1050,6 +1050,7 @@
 extern struct vfsmount *kern_mount(struct file_system_type *);
 extern int may_umount(struct vfsmount *);
 extern long do_mount(char *, char *, char *, unsigned long, void *);
+struct vfsmount *do_kern_mount(const char *fstype, int flags, char *name, void *data);
 extern void umount_tree(struct vfsmount *);
 
 #define kern_umount mntput
Index: linux-ion/mm/memory.c
===================================================================
--- linux-ion.orig/mm/memory.c	2004-07-28 14:34:57.000000000 -0700
+++ linux-ion/mm/memory.c	2004-09-27 15:05:56.000000000 -0700
@@ -401,6 +401,7 @@
 		mm->rss = 0;
 	spin_unlock(&mm->page_table_lock);
 }
+EXPORT_SYMBOL(zap_page_range);
 
 /*
  * Do a quick page-table lookup for a single page. 
