Index: linux-2.6/fs/jbd/journal.c
===================================================================
--- linux-2.6.orig/fs/jbd2/journal.c	2006-07-15 16:13:50.000000000 +0800
+++ linux-2.6/fs/jbd2/journal.c	2006-07-15 16:22:04.000000000 +0800
@@ -74,6 +74,7 @@ EXPORT_SYMBOL(journal_abort);
 	spin_unlock(&journal->j_state_lock);
 	return ret;
 }
+EXPORT_SYMBOL(jbd2_log_start_commit);
 
 /*
  * Force and wait upon a commit if the calling process is not within
Index: linux-2.6/kernel/sys.c
===================================================================
--- linux-2.6.orig/security/security.c
+++ linux-2.6/security/security.c
@@ -74,6 +74,7 @@
 		return 0;
 	return security_ops->inode_unlink(dir, dentry, mnt);
 }
+EXPORT_SYMBOL(security_inode_unlink);
 
 int security_inode_symlink(struct inode *dir, struct dentry *dentry,
 			   struct vfsmount *mnt, const char *old_name)
