Index: linux-2.6.0-test5/net/core/netpoll.c
===================================================================
--- linux-2.6.0-test5.orig/net/core/netpoll.c	2003-09-26 15:42:10.000000000 +0800
+++ linux-2.6.0-test5/net/core/netpoll.c	2003-09-26 15:42:32.000000000 +0800
@@ -630,3 +630,11 @@
 {
 	trapped = trap;
 }
+EXPORT_SYMBOL(netpoll_set_trap);
+EXPORT_SYMBOL(netpoll_trap);
+EXPORT_SYMBOL(netpoll_parse_options);
+EXPORT_SYMBOL(netpoll_setup);
+EXPORT_SYMBOL(netpoll_cleanup);
+EXPORT_SYMBOL(netpoll_send_skb);
+EXPORT_SYMBOL(netpoll_send_udp);
+EXPORT_SYMBOL(netpoll_poll);
