Index: linux-2.6.3/fs/jbd/journal.c
===================================================================
--- linux-2.6.3.orig/fs/jbd/journal.c	2004-01-08 22:59:10.000000000 -0800
+++ linux-2.6.3/fs/jbd/journal.c	2004-02-23 20:09:34.000000000 -0800
@@ -71,6 +71,7 @@
 EXPORT_SYMBOL(journal_errno);
 EXPORT_SYMBOL(journal_ack_err);
 EXPORT_SYMBOL(journal_clear_err);
+EXPORT_SYMBOL(log_start_commit);
 EXPORT_SYMBOL(log_wait_commit);
 EXPORT_SYMBOL(journal_start_commit);
 EXPORT_SYMBOL(journal_wipe);
