--- linux/arch/i386/kernel/i386_ksyms.c.orig	2003-11-28 11:26:19.000000000 -0700
+++ linux/arch/i386/kernel/i386_ksyms.c	2004-08-13 11:31:33.000000000 -0600
@@ -131,6 +131,7 @@
 EXPORT_SYMBOL(cpu_data);
 EXPORT_SYMBOL(kernel_flag_cacheline);
 EXPORT_SYMBOL(smp_num_cpus);
+EXPORT_SYMBOL(smp_num_siblings);
 EXPORT_SYMBOL(cpu_online_map);
 EXPORT_SYMBOL_NOVERS(__write_lock_failed);
 EXPORT_SYMBOL_NOVERS(__read_lock_failed);
