Index: linux-stage/include/linux/ext3_fs_sb.h
===================================================================
--- linux-stage.orig/include/linux/ext3_fs_sb.h	2004-11-03 14:16:26.059485670 -0500
+++ linux-stage/include/linux/ext3_fs_sb.h	2004-11-03 14:20:30.761493072 -0500
@@ -19,9 +19,12 @@
 #ifdef __KERNEL__
 #include <linux/timer.h>
 #include <linux/wait.h>
+#ifndef EXT_INCLUDE
+#define EXT_INCLUDE
 #include <linux/blockgroup_lock.h>
 #include <linux/percpu_counter.h>
 #endif
+#endif
 #include <linux/rbtree.h>
 
 /*
