Index: linux-2.4.24-l36mmap/mm/memory.c
===================================================================
--- linux-2.4.24-l36mmap.orig/mm/memory.c	2004-05-27 17:44:13.000000000 -0700
+++ linux-2.4.24-l36mmap/mm/memory.c	2004-05-27 17:45:07.000000000 -0700
@@ -411,6 +411,7 @@
 		mm->rss = 0;
 	spin_unlock(&mm->page_table_lock);
 }
+EXPORT_SYMBOL_GPL(zap_page_range);
 
 /*
  * Do a quick page-table lookup for a single page. 
