Index: linux-2.6/fs/open.c
===================================================================
--- linux-2.6.orig/fs/open.c	2006-07-15 16:10:37.000000000 +0800
+++ linux-2.6/fs/open.c	2006-07-15 16:22:04.000000000 +0800
@@ -808,7 +808,6 @@ asmlinkage long sys_lchown(const char __
 	return error;
 }
 
-
 asmlinkage long sys_fchown(unsigned int fd, uid_t user, gid_t group)
 {
 	struct file * file;
Index: linux-2.6/fs/jbd/journal.c
===================================================================
--- linux-2.6.orig/fs/jbd/journal.c	2006-07-15 16:13:50.000000000 +0800
+++ linux-2.6/fs/jbd/journal.c	2006-07-15 16:22:04.000000000 +0800
@@ -74,6 +74,7 @@ EXPORT_SYMBOL(journal_abort);
 EXPORT_SYMBOL(journal_errno);
 EXPORT_SYMBOL(journal_ack_err);
 EXPORT_SYMBOL(journal_clear_err);
+EXPORT_SYMBOL(log_start_commit);
 EXPORT_SYMBOL(log_wait_commit);
 EXPORT_SYMBOL(journal_start_commit);
 EXPORT_SYMBOL(journal_force_commit_nested);
