 include/linux/ext3_fs.h |    1 +
 1 files changed, 1 insertion(+)

--- linux-2.4.18/include/linux/ext3_fs.h~ext-2.4-patch-5	2003-08-29 16:53:18.000000000 +0400
+++ linux-2.4.18-alexey/include/linux/ext3_fs.h	2003-09-01 11:50:37.000000000 +0400
@@ -344,6 +344,7 @@ struct ext3_inode {
   #define EXT3_MOUNT_WRITEBACK_DATA	0x0C00	/* No data ordering */
 #define EXT3_MOUNT_UPDATE_JOURNAL	0x1000	/* Update the journal format */
 #define EXT3_MOUNT_NO_UID32		0x2000  /* Disable 32-bit UIDs */
+#define EXT3_MOUNT_INDEX		0x4000  /* Enable directory index */
 
 /* Compatibility, for having both ext2_fs.h and ext3_fs.h included at once */
 #ifndef _LINUX_EXT2_FS_H

_
