Index: linux-2.6/include/linux/dcache.h
===================================================================
--- linux-2.6.orig/include/linux/dcache.h	2006-07-15 16:11:52.000000000 +0800
+++ linux-2.6/include/linux/dcache.h	2006-07-15 16:12:04.000000000 +0800
@@ -24,6 +24,11 @@ struct vfsmount;
 
 #define IS_ROOT(x) ((x) == (x)->d_parent)
 
+struct dentry_params {
+	unsigned long   p_inum;
+	void            *p_ptr;
+};
+
 /*
  * "quick string" -- eases parameter passing, but more importantly
  * saves "metadata" about the string (ie length and the hash).
