Index: linux-2.6.15/include/linux/ext3_fs_i.h
===================================================================
--- linux-2.6.15.orig/include/linux/ext3_fs_i.h	2006-02-24 15:41:30.000000000 +0300
+++ linux-2.6.15/include/linux/ext3_fs_i.h	2006-02-24 15:41:31.000000000 +0300
@@ -135,6 +135,8 @@ struct ext3_inode_info {
 	struct inode vfs_inode;
 
 	__u32 i_cached_extent[4];
+
+	void *i_filterdata;
 };
 
 #endif	/* _LINUX_EXT3_FS_I */
Index: linux-2.6.15/fs/ext3/super.c
===================================================================
--- linux-2.6.15.orig/fs/ext3/super.c	2006-02-24 15:41:30.000000000 +0300
+++ linux-2.6.15/fs/ext3/super.c	2006-02-24 15:42:02.000000000 +0300
@@ -459,6 +459,7 @@ static struct inode *ext3_alloc_inode(st
 	ei->vfs_inode.i_version = 1;
 	
 	memset(&ei->i_cached_extent, 0, sizeof(ei->i_cached_extent));
+	ei->i_filterdata = NULL;
 	return &ei->vfs_inode;
 }
 
