 fs/ext3/ext3-exports.c |    5 +++++
 1 files changed, 5 insertions(+)

--- linux/fs/ext3/ext3-exports.c~ext3-error-export	Mon Jul 14 19:30:50 2003
+++ linux-mmonroe/fs/ext3/ext3-exports.c	Mon Jul 14 19:32:37 2003
@@ -21,3 +21,8 @@ EXPORT_SYMBOL(ext3_xattr_list);
 EXPORT_SYMBOL(ext3_xattr_set);
 EXPORT_SYMBOL(ext3_prep_san_write);
 EXPORT_SYMBOL(ext3_map_inode_page);
+
+EXPORT_SYMBOL(ext3_abort);
+EXPORT_SYMBOL(ext3_decode_error);
+EXPORT_SYMBOL(__ext3_std_error);
+

_
