Index: linux-2.6.10/fs/ext3/super.c
===================================================================
--- linux-2.6.10.orig/fs/ext3/super.c	2005-04-06 19:38:35.000000000 +0400
+++ linux-2.6.10/fs/ext3/super.c	2005-06-23 21:25:44.000000000 +0400
@@ -1313,7 +1313,9 @@
 	sbi->s_resuid = le16_to_cpu(es->s_def_resuid);
 	sbi->s_resgid = le16_to_cpu(es->s_def_resgid);
 
+#if 0
 	set_opt(sbi->s_mount_opt, RESERVATION);
+#endif
 
 	if (!parse_options ((char *) data, sb, &journal_inum, NULL, 0))
 		goto failed_mount;
