Index: linux-2.4.19.SuSE/fs/ext3/namei.c
===================================================================
--- linux-2.4.19.SuSE.orig/fs/ext3/namei.c	Sun Nov 16 01:14:50 2003
+++ linux-2.4.19.SuSE/fs/ext3/namei.c	Sun Nov 16 01:18:04 2003
@@ -1751,8 +1751,8 @@
 	struct super_block *sb = inode->i_sb;
 	struct ext3_iloc iloc;
 	int err = 0, rc;
-	
-	lock_super(sb);
+
+	down(&EXT3_SB(sb)->s_orphan_lock);
 	if (!list_empty(&EXT3_I(inode)->i_orphan))
 		goto out_unlock;
 
@@ -1800,7 +1800,7 @@
 	jbd_debug(4, "orphan inode %ld will point to %d\n",
 			inode->i_ino, NEXT_ORPHAN(inode));
 out_unlock:
-	unlock_super(sb);
+	up(&EXT3_SB(sb)->s_orphan_lock);
 	ext3_std_error(inode->i_sb, err);
 	return err;
 }
@@ -1813,20 +1813,19 @@
 {
 	struct list_head *prev;
 	struct ext3_inode_info *ei = EXT3_I(inode);
-	struct ext3_sb_info *sbi;
+	struct ext3_sb_info *sbi = EXT3_SB(inode->i_sb);
 	unsigned long ino_next;
 	struct ext3_iloc iloc;
 	int err = 0;
 
-	lock_super(inode->i_sb);
+	down(&sbi->s_orphan_lock);
 	if (list_empty(&ei->i_orphan)) {
-		unlock_super(inode->i_sb);
+		up(&sbi->s_orphan_lock);
 		return 0;
 	}
 
 	ino_next = NEXT_ORPHAN(inode);
 	prev = ei->i_orphan.prev;
-	sbi = EXT3_SB(inode->i_sb);
 
 	jbd_debug(4, "remove inode %lu from orphan list\n", inode->i_ino);
 
@@ -1872,10 +1871,10 @@
 	if (err)
 		goto out_brelse;
 
-out_err: 
+out_err:
 	ext3_std_error(inode->i_sb, err);
 out:
-	unlock_super(inode->i_sb);
+	up(&sbi->s_orphan_lock);
 	return err;
 
 out_brelse:
Index: linux-2.4.19.SuSE/fs/ext3/super.c
===================================================================
--- linux-2.4.19.SuSE.orig/fs/ext3/super.c	Sun Nov 16 00:40:59 2003
+++ linux-2.4.19.SuSE/fs/ext3/super.c	Sun Nov 16 01:18:04 2003
@@ -1182,6 +1182,7 @@
 	 */
 	sb->s_op = &ext3_sops;
 	INIT_LIST_HEAD(&sbi->s_orphan); /* unlinked but open files */
+	sema_init(&sbi->s_orphan_lock, 1);
 
 	sb->s_root = 0;
 
Index: linux-2.4.19.SuSE/include/linux/ext3_fs_sb.h
===================================================================
--- linux-2.4.19.SuSE.orig/include/linux/ext3_fs_sb.h	Sat Nov 15 23:58:28 2003
+++ linux-2.4.19.SuSE/include/linux/ext3_fs_sb.h	Sun Nov 16 01:18:41 2003
@@ -69,6 +69,7 @@
 	struct inode * s_journal_inode;
 	struct journal_s * s_journal;
 	struct list_head s_orphan;
+ 	struct semaphore s_orphan_lock;
 	struct block_device *journal_bdev;
 #ifdef CONFIG_JBD_DEBUG
 	struct timer_list turn_ro_timer;	/* For turning read-only (crash simulation) */
