Index: linux-2.6.0/fs/jbd/transaction.c
===================================================================
--- linux-2.6.0.orig/fs/jbd/transaction.c	2004-01-05 16:18:59.000000000 +0300
+++ linux-2.6.0/fs/jbd/transaction.c	2004-01-05 16:19:48.000000000 +0300
@@ -85,7 +85,7 @@
 	int needed;
 	int nblocks = handle->h_buffer_credits;
 	transaction_t *new_transaction = NULL;
-	int ret;
+	int ret = 0;
 
 	if (nblocks > journal->j_max_transaction_buffers) {
 		printk(KERN_ERR "JBD: %s wants too many credits (%d > %d)\n",
@@ -231,7 +231,7 @@
 out:
 	if (new_transaction)
 		kfree(new_transaction);
-	return 0;
+	return ret;
 }
 
 /* Allocate a new handle.  This should probably be in a slab... */
