--- linux/fs/ext3/namei.c.orig	Fri Mar 14 14:11:58 2003
+++ linux/fs/ext3/namei.c	Fri Mar 14 14:39:48 2003
@@ -1406,8 +1409,8 @@
 	struct super_block *sb = inode->i_sb;
 	struct ext3_iloc iloc;
 	int err = 0, rc;
-	
-	lock_super(sb);
+
+	down(&EXT3_SB(sb)->s_orphan_lock);
 	if (!list_empty(&EXT3_I(inode)->i_orphan))
 		goto out_unlock;
 
@@ -1455,7 +1458,7 @@
 	jbd_debug(4, "orphan inode %ld will point to %d\n",
 			inode->i_ino, NEXT_ORPHAN(inode));
 out_unlock:
-	unlock_super(sb);
+	up(&EXT3_SB(sb)->s_orphan_lock);
 	ext3_std_error(inode->i_sb, err);
 	return err;
 }
@@ -1468,20 +1471,19 @@
 {
 	struct list_head *prev;
 	struct ext3_inode_info *ei = EXT3_I(inode);
-	struct ext3_sb_info *sbi;
+	struct ext3_sb_info *sbi = EXT3_SB(inode->i_sb);
 	unsigned long ino_next;
 	struct ext3_iloc iloc;
 	int err = 0;
 
-	lock_super(inode->i_sb);
+	down(&sbi->s_orphan_lock);
 	if (list_empty(&ei->i_orphan)) {
-		unlock_super(inode->i_sb);
+		up(&sbi->s_orphan_lock);
 		return 0;
 	}
 
 	ino_next = NEXT_ORPHAN(inode);
 	prev = ei->i_orphan.prev;
-	sbi = EXT3_SB(inode->i_sb);
 
 	jbd_debug(4, "remove inode %lu from orphan list\n", inode->i_ino);
 
@@ -1525,10 +1527,10 @@
 	if (err)
 		goto out_brelse;
 
-out_err: 
+out_err:
 	ext3_std_error(inode->i_sb, err);
 out:
-	unlock_super(inode->i_sb);
+	up(&sbi->s_orphan_lock);
 	return err;
 
 out_brelse:
--- linux/fs/ext3/super.c.orig	Fri Mar 14 14:11:58 2003
+++ linux/fs/ext3/super.c	Fri Mar 14 14:36:00 2003
@@ -1134,6 +1314,7 @@
 	 */
 	sb->s_op = &ext3_sops;
 	INIT_LIST_HEAD(&sbi->s_orphan); /* unlinked but open files */
+	sema_init(&sbi->s_orphan_lock, 1);
 
 	sb->s_root = 0;
 
--- linux/include/linux/ext3_fs_sb.h.orig	Tue Feb 11 16:34:33 2003
+++ linux/include/linux/ext3_fs_sb.h	Fri Mar 14 14:30:11 2003
@@ -67,6 +69,7 @@
 	struct inode * s_journal_inode;
 	struct journal_s * s_journal;
 	struct list_head s_orphan;
+	struct semaphore s_orphan_lock;
 	unsigned long s_commit_interval;
 	struct block_device *journal_bdev;
 #ifdef CONFIG_JBD_DEBUG
