Index: 57chaos/fs/ext3/inode.c
===================================================================
--- 57chaos.orig/fs/ext3/inode.c	2004-06-21 14:15:31.000000000 -0700
+++ 57chaos/fs/ext3/inode.c	2004-06-21 14:19:27.000000000 -0700
@@ -2270,6 +2270,12 @@ void ext3_truncate_thread(struct inode *
 
 	memcpy(nei->i_data, oei->i_data, sizeof(nei->i_data));
 	memset(oei->i_data, 0, sizeof(oei->i_data));
+	if (EXT3_I(old_inode)->i_flags & EXT3_EXTENTS_FL) {
+		EXT3_I(new_inode)->i_flags |= EXT3_EXTENTS_FL;
+		ext3_extents_initialize_blockmap(handle, old_inode);
+	} else {
+		EXT3_I(new_inode)->i_flags &= ~EXT3_EXTENTS_FL;
+	}
 
 	nei->i_disksize = oei->i_disksize;
 	nei->i_state |= EXT3_STATE_DELETE;
@@ -2522,6 +2526,13 @@ void ext3_read_inode(struct inode * inod
 	else
 		EXT3_I(inode)->i_extra_isize = 0;
 
+	if (EXT3_I(inode)->i_flags & EXT3_EXTENTS_FL) {
+		inode->u.ext3_i.i_cached_extent[0] = 0;
+		inode->u.ext3_i.i_cached_extent[1] = 0;
+		inode->u.ext3_i.i_cached_extent[2] = 0;
+		inode->u.ext3_i.i_cached_extent[3] = 0;
+	}
+
 	if (S_ISREG(inode->i_mode)) {
 		inode->i_op = &ext3_file_inode_operations;
 		inode->i_fop = &ext3_file_operations;
