
Index: linux-2.4.21-chaos/fs/ext3/inode.c
===================================================================
--- linux-2.4.21-chaos.orig/fs/ext3/inode.c	2003-12-12 16:19:13.000000000 +0300
+++ linux-2.4.21-chaos/fs/ext3/inode.c	2003-12-12 16:19:16.000000000 +0300
@@ -3154,7 +3154,7 @@
 	/* alloc blocks one by one */
 	for (i = 0; i < nblocks; i++) {
 		ret = ext3_get_block_handle(handle, inode, blocks[i],
-						&bh_tmp, 1);
+						&bh_tmp, 1, 1);
 		if (ret)
 			break;
 
@@ -3229,7 +3229,7 @@
                 if (blocks[i] != 0)
                         continue;
 
-                rc = ext3_get_block_handle(handle, inode, iblock, &bh, 1);
+                rc = ext3_get_block_handle(handle, inode, iblock, &bh, 1, 1);
                 if (rc) {
                         printk(KERN_INFO "ext3_map_inode_page: error %d "
                                "allocating block %ld\n", rc, iblock);
