Index: linux-2.4.21-suse/fs/ext3/namei.c
===================================================================
--- linux-2.4.21-suse.orig/fs/ext3/namei.c	2003-10-30 02:17:22.000000000 +0300
+++ linux-2.4.21-suse/fs/ext3/namei.c	2003-10-30 02:20:53.000000000 +0300
@@ -1747,8 +1747,8 @@
 	struct super_block *sb = inode->i_sb;
 	struct ext3_iloc iloc;
 	int err = 0, rc;
-	
-	lock_super(sb);
+
+	down(&EXT3_SB(sb)->s_orphan_lock);
 	if (!list_empty(&EXT3_I(inode)->i_orphan))
 		goto out_unlock;
 
@@ -1796,7 +1796,7 @@
 	jbd_debug(4, "orphan inode %ld will point to %d\n",
 			inode->i_ino, NEXT_ORPHAN(inode));
 out_unlock:
-	unlock_super(sb);
+	up(&EXT3_SB(sb)->s_orphan_lock);
 	ext3_std_error(inode->i_sb, err);
 	return err;
 }
@@ -1808,20 +1808,19 @@
 int ext3_orphan_del(handle_t *handle, struct inode *inode)
 {
 	struct list_head *prev;
-	struct ext3_sb_info *sbi;
+	struct ext3_sb_info *sbi = EXT3_SB(inode->i_sb);
 	unsigned long ino_next;
 	struct ext3_iloc iloc;
 	int err = 0;
 
-	lock_super(inode->i_sb);
+	down(&sbi->s_orphan_lock);
 	if (list_empty(&inode->u.ext3_i.i_orphan)) {
-		unlock_super(inode->i_sb);
+		up(&sbi->s_orphan_lock);
 		return 0;
 	}
 
 	ino_next = NEXT_ORPHAN(inode);
 	prev = inode->u.ext3_i.i_orphan.prev;
-	sbi = EXT3_SB(inode->i_sb);
 
 	jbd_debug(4, "remove inode %lu from orphan list\n", inode->i_ino);
 
@@ -1870,7 +1869,7 @@
 out_err:
 	ext3_std_error(inode->i_sb, err);
 out:
-	unlock_super(inode->i_sb);
+	up(&sbi->s_orphan_lock);
 	return err;
 
 out_brelse:
Index: linux-2.4.21-suse/fs/ext3/super.c
===================================================================
--- linux-2.4.21-suse.orig/fs/ext3/super.c	2003-10-30 02:17:22.000000000 +0300
+++ linux-2.4.21-suse/fs/ext3/super.c	2003-10-30 02:17:22.000000000 +0300
@@ -1151,6 +1151,7 @@
 	 */
 	sb->s_op = &ext3_sops;
 	INIT_LIST_HEAD(&sbi->s_orphan); /* unlinked but open files */
+	sema_init(&sbi->s_orphan_lock, 1);
 
 	sb->s_root = 0;
 
Index: linux-2.4.21-suse/include/linux/ext3_fs_sb.h
===================================================================
--- linux-2.4.21-suse.orig/include/linux/ext3_fs_sb.h	2003-10-30 02:17:22.000000000 +0300
+++ linux-2.4.21-suse/include/linux/ext3_fs_sb.h	2003-10-30 02:17:22.000000000 +0300
@@ -69,6 +69,7 @@
 	struct inode * s_journal_inode;
 	struct journal_s * s_journal;
 	struct list_head s_orphan;
+	struct semaphore s_orphan_lock;
 	unsigned long s_commit_interval;
 	struct block_device *journal_bdev;
 #ifdef CONFIG_JBD_DEBUG
