 fs/ext3/inode.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

--- linux-2.4.20/fs/ext3/inode.c~ext3-largefile	2003-04-08 23:35:36.000000000 -0600
+++ linux-2.4.20-braam/fs/ext3/inode.c	2003-04-08 23:35:36.000000000 -0600
@@ -2562,7 +2562,7 @@ void ext3_dirty_inode(struct inode *inod
 	handle_t *handle;
 
 	lock_kernel();
-	handle = ext3_journal_start(inode, 1);
+	handle = ext3_journal_start(inode, 2);
 	if (IS_ERR(handle))
 		goto out;
 	if (current_handle &&

_
