
--- linux-2.4.18/fs/ext3/super.c~ext3-no-write-super	Mon Jul 28 14:13:05 2003
+++ linux-2.4.18-alexey/fs/ext3/super.c	Mon Jul 28 16:14:11 2003
@@ -1818,7 +1818,10 @@ void ext3_write_super (struct super_bloc
 	if (down_trylock(&sb->s_lock) == 0)
 		BUG();
 	sb->s_dirt = 0;
+#if 0
+	/* we really don't need this, jbd makes periodical commits by itself */
 	log_start_commit(EXT3_SB(sb)->s_journal, NULL);
+#endif
 }
 
 static int ext3_sync_fs(struct super_block *sb)

_
