 fs/ext3/super.c |    1 -
 1 files changed, 1 deletion(-)

--- linux-2.4.18-chaos/fs/ext3/super.c~ext3-no-write-super-chaos	2003-08-24 21:34:53.000000000 +0400
+++ linux-2.4.18-chaos-alexey/fs/ext3/super.c	2003-08-24 21:40:47.000000000 +0400
@@ -1818,7 +1818,6 @@ void ext3_write_super (struct super_bloc
 	if (down_trylock(&sb->s_lock) == 0)
 		BUG();
 	sb->s_dirt = 0;
-	log_start_commit(EXT3_SB(sb)->s_journal, NULL);
 }
 
 static int ext3_sync_fs(struct super_block *sb)

_
