diff -rupN linux-2.4.24.orig/fs/ext3/namei.c linux-2.4.24/fs/ext3/namei.c
--- linux-2.4.24.orig/fs/ext3/namei.c	Tue Oct  5 10:34:28 2004
+++ linux-2.4.24/fs/ext3/namei.c	Tue Oct  5 10:43:12 2004
@@ -1112,6 +1112,7 @@ static struct dentry *ext3_lookup(struct
 			dentry->d_flags |= DCACHE_CROSS_REF;
 			dentry->d_generation = mds[1];
 			dentry->d_mdsnum = mds[0];
+			dentry->d_fid = mds[2];
 			dentry->d_inum = ino;
 			d_add(dentry, NULL);
 			return NULL;
@@ -1334,7 +1335,8 @@ static int add_dirent_to_buf(handle_t *h
 	if (EXT3_HAS_INCOMPAT_FEATURE(sb, EXT3_FEATURE_INCOMPAT_MDSNUM)
 			&& (dentry->d_flags & DCACHE_CROSS_REF)
 			&& (dentry->d_mdsnum != EXT3_SB(sb)->s_mdsnum))
-		reclen += 8; /* we need space to store mds num */
+		reclen += 12; /* we need space to store mds num, 
+                               * inode num and fid num. */
 	if (!de) {
 		de = (struct ext3_dir_entry_2 *)bh->b_data;
 		top = bh->b_data + dir->i_sb->s_blocksize - reclen;
@@ -1387,6 +1389,7 @@ static int add_dirent_to_buf(handle_t *h
 			mds = (__u32 *)((char *)de + EXT3_DIR_REC_LEN(namelen));
 			mds[0] = cpu_to_le32(dentry->d_mdsnum);
 			mds[1] = cpu_to_le32(dentry->d_generation);
+			mds[2] = cpu_to_le32(dentry->d_fid);
 			de->inode = cpu_to_le32(dentry->d_inum);
 			de->file_type = 128;
 		} else {
diff -rupN linux-2.4.24.orig/include/linux/dcache.h linux-2.4.24/include/linux/dcache.h
--- linux-2.4.24.orig/include/linux/dcache.h	Tue Oct  5 10:34:28 2004
+++ linux-2.4.24/include/linux/dcache.h	Tue Oct  5 10:43:45 2004
@@ -123,6 +123,7 @@ struct dentry {
 	unsigned d_inum;		/* for cross-fs references (Lustre) */
 	unsigned d_mdsnum;		/* for cross-fs references (Lustre) */
 	unsigned d_generation;		/* for cross-fs references (Lustre) */
+        unsigned d_fid;                 /* for cross-fs references (Lustre) */
 	struct dentry * d_parent;	/* parent directory */
 	struct list_head d_hash;	/* lookup hash list */
 	struct list_head d_lru;		/* d_count = 0 LRU list */
diff -rupN linux-2.4.24.orig/include/linux/ext3_fs.h linux-2.4.24/include/linux/ext3_fs.h
--- linux-2.4.24.orig/include/linux/ext3_fs.h	Tue Oct  5 10:34:28 2004
+++ linux-2.4.24/include/linux/ext3_fs.h	Tue Oct  5 10:44:29 2004
@@ -589,7 +589,7 @@ struct ext3_dir_entry_2 {
 #define EXT3_DIR_REC_LEN(name_len)	(((name_len) + 8 + EXT3_DIR_ROUND) & \
 					 ~EXT3_DIR_ROUND)
 #define EXT3_DIR_REC_LEN_DE(de)		(EXT3_DIR_REC_LEN((de)->name_len) + \
-					(((de)->file_type & 128) ? 8 : 0))
+					(((de)->file_type & 128) ? 12 : 0))
 
 /*
  * Hash Tree Directory indexing
