Index: uml/include/linux/ext3_fs.h
===================================================================
--- uml.orig/include/linux/ext3_fs.h	2004-12-24 16:34:58.000000000 -0500
+++ uml/include/linux/ext3_fs.h	2005-04-15 00:10:56.220649968 -0400
@@ -353,8 +353,8 @@
 #define EXT3_MOUNT_NO_UID32		0x02000  /* Disable 32-bit UIDs */
 #define EXT3_MOUNT_XATTR_USER		0x04000	/* Extended user attributes */
 #define EXT3_MOUNT_POSIX_ACL		0x08000	/* POSIX Access Control Lists */
-#define EXT3_MOUNT_RESERVATION		0x10000	/* Preallocation */
-#define EXT3_MOUNT_BARRIER		0x20000 /* Use block barriers */
+#define EXT3_MOUNT_BARRIER		0x10000 /* Use block barriers */
+#define EXT3_MOUNT_RESERVATION		0x20000	/* Preallocation */
 
 /* Compatibility, for having both ext2_fs.h and ext3_fs.h included at once */
 #ifndef _LINUX_EXT2_FS_H
