Index: linux-2.4.20-30.9/fs/ext3/inode.c
===================================================================
--- linux-2.4.20-30.9.orig/fs/ext3/inode.c	2004-05-18 13:55:49.000000000 -0700
+++ linux-2.4.20-30.9/fs/ext3/inode.c	2004-05-18 13:59:46.000000000 -0700
@@ -2217,6 +2217,10 @@
 
 	memcpy(nei->i_data, oei->i_data, sizeof(nei->i_data));
 	memset(oei->i_data, 0, sizeof(oei->i_data));
+ 	if (EXT3_I(old_inode)->i_flags & EXT3_EXTENTS_FL) {
+ 		EXT3_I(new_inode)->i_flags |= EXT3_EXTENTS_FL;
+  		ext3_extents_initialize_blockmap(handle, old_inode);
+ 	}
 
 	nei->i_disksize = oei->i_disksize;
 	nei->i_state |= EXT3_STATE_DELETE;
@@ -2447,6 +2451,12 @@
 
 	brelse (iloc.bh);
 
+	if (EXT3_I(inode)->i_flags & EXT3_EXTENTS_FL) {
+		inode->u.ext3_i.i_cached_extent[0] = 0;
+		inode->u.ext3_i.i_cached_extent[1] = 0;
+		inode->u.ext3_i.i_cached_extent[2] = 0;
+	}
+
 	if (S_ISREG(inode->i_mode)) {
 		inode->i_op = &ext3_file_inode_operations;
 		inode->i_fop = &ext3_file_operations;
