Index: linux-2.6.0/fs/ext3/super.c
===================================================================
--- linux-2.6.0.orig/fs/ext3/super.c	2003-12-31 16:00:58.000000000 +0300
+++ linux-2.6.0/fs/ext3/super.c	2004-01-03 11:25:32.000000000 +0300
@@ -1068,6 +1068,7 @@
 	sbi->s_mount_opt = 0;
 	sbi->s_resuid = EXT3_DEF_RESUID;
 	sbi->s_resgid = EXT3_DEF_RESGID;
+	sbi->s_next_generation = jiffies;
 
 	blocksize = sb_min_blocksize(sb, EXT3_MIN_BLOCK_SIZE);
 	if (!blocksize) {
