--- linux-2.4.17/fs/ext3/super.c.orig	Fri Dec 21 10:41:55 2001
+++ linux-2.4.17/fs/ext3/super.c	Fri Mar 22 11:00:41 2002
@@ -1742,7 +1742,7 @@
 	unregister_filesystem(&ext3_fs_type);
 }
 
-EXPORT_NO_SYMBOLS;
+EXPORT_SYMBOL(ext3_bread);
 
 MODULE_AUTHOR("Remy Card, Stephen Tweedie, Andrew Morton, Andreas Dilger, Theodore Ts'o and others");
 MODULE_DESCRIPTION("Second Extended Filesystem with journaling extensions");
