 fs/ext3/super.c |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

--- linux-2.4.18-p4smp/fs/ext3/super.c~extN-misc-fixup	2003-07-21 23:07:50.000000000 -0600
+++ linux-2.4.18-p4smp-braam/fs/ext3/super.c	2003-07-21 23:08:06.000000000 -0600
@@ -1578,10 +1578,10 @@ static journal_t *ext3_get_dev_journal(s
 		printk(KERN_ERR "EXT3-fs: I/O error on journal device\n");
 		goto out_journal;
 	}
-	if (ntohl(journal->j_superblock->s_nr_users) != 1) {
+	if (be32_to_cpu(journal->j_superblock->s_nr_users) != 1) {
 		printk(KERN_ERR "EXT3-fs: External journal has more than one "
 					"user (unsupported) - %d\n",
-			ntohl(journal->j_superblock->s_nr_users));
+			be32_to_cpu(journal->j_superblock->s_nr_users));
 		goto out_journal;
 	}
 	EXT3_SB(sb)->journal_bdev = bdev;

_
