Index: linux-2.6.10/include/linux/namei.h
===================================================================
--- linux-2.6.10.orig/include/linux/namei.h	2005-04-06 09:38:35.000000000 -0600
+++ linux-2.6.10/include/linux/namei.h	2006-01-03 15:32:11.000000000 -0700
@@ -46,6 +46,7 @@
 #define LOOKUP_PARENT		16
 #define LOOKUP_NOALT		32
 #define LOOKUP_ATOMIC		64
+#define LOOKUP_REVAL		128
 
 /*
  * Intent data
