--- kernel-2.4.20-6chaos_18_7/include/linux/module.h~gpl-header	Wed Jul 16 03:54:37 2003
+++ kernel-2.4.20-6chaos_18_7/include/linux/module.h	Wed Jul 16 03:55:45 2003
@@ -288,7 +288,7 @@
 "license=" license
 
 #define GPL_HEADER() \
-static const char cpyright="This software may be freely redistributed under the terms of the GNU General Public License.";
+static const char cpyright[]="This software may be freely redistributed under the terms of the GNU General Public License.";
 
 /* Define the module variable, and usage macros.  */
 extern struct module __this_module;
