Index: linux-2.6.18.8/include/linux/fs.h
===================================================================
--- linux-2.6.18.8.orig/include/linux/fs.h	2007-06-05 12:55:19.000000000 +0200
+++ linux-2.6.18.8/include/linux/fs.h	2007-06-05 12:55:44.000000000 +0200
@@ -533,6 +533,7 @@ struct inode {
 	struct block_device	*i_bdev;
 	struct cdev		*i_cdev;
 	int			i_cindex;
+	void                    *i_filterdata;
 
 	__u32			i_generation;
 
