Index: linux-2.6.18-128.1.14/include/linux/fs.h
===================================================================
--- linux-2.6.18-128.1.14.orig/include/linux/fs.h	2009-06-19 12:33:13.000000000 -0600
+++ linux-2.6.18-128.1.14/include/linux/fs.h	2009-06-19 12:33:26.000000000 -0600
@@ -580,6 +580,7 @@
 		struct cdev		*i_cdev;
 	};
 	int			i_cindex;
+	void                    *i_filterdata;
 
 	__u32			i_generation;
 
