
diff -puN fs/fs-writeback.c~inode-protection-from-pdflush fs/fs-writeback.c
--- linux-2.5.73/fs/fs-writeback.c~inode-protection-from-pdflush	Mon Jul  7 01:10:17 2003
+++ linux-2.5.73-alexey/fs/fs-writeback.c	Mon Jul  7 01:10:17 2003
@@ -198,6 +198,11 @@ static void
 __writeback_single_inode(struct inode *inode,
 			struct writeback_control *wbc)
 {
+	if (inode->i_flags & I_SKIP_PDFLUSH) {
+		list_move(&inode->i_list, &inode->i_sb->s_dirty);
+		return;
+	}
+
 	if ((wbc->sync_mode != WB_SYNC_ALL) && (inode->i_state & I_LOCK)) {
 		list_move(&inode->i_list, &inode->i_sb->s_dirty);
 		return;
diff -puN include/linux/fs.h~inode-protection-from-pdflush include/linux/fs.h
--- linux-2.5.73/include/linux/fs.h~inode-protection-from-pdflush	Mon Jul  7 01:22:16 2003
+++ linux-2.5.73-alexey/include/linux/fs.h	Mon Jul  7 01:23:18 2003
@@ -816,6 +816,7 @@ struct super_operations {
 #define I_FREEING		16
 #define I_CLEAR			32
 #define I_NEW			64
+#define I_SKIP_PDFLUSH		1024 /* inode will be out of pdflush's scope */
 
 #define I_DIRTY (I_DIRTY_SYNC | I_DIRTY_DATASYNC | I_DIRTY_PAGES)
 

_
