--- linux-chaos/fs/inode.c.b_io_export	Wed Jan 29 16:56:15 2003
+++ linux-chaos/fs/inode.c	Wed Jan 29 16:56:27 2003
@@ -66,7 +66,8 @@
  * NOTE! You also have to own the lock if you change
  * the i_state of an inode while it is in use..
  */
-static spinlock_t inode_lock = SPIN_LOCK_UNLOCKED;
+spinlock_t inode_lock = SPIN_LOCK_UNLOCKED;
+EXPORT_SYMBOL(inode_lock);
 
 /*
  * Statistics gathering..
--- linux-chaos/fs/Makefile.b_io_export	Wed Jan 29 16:56:45 2003
+++ linux-chaos/fs/Makefile	Wed Jan 29 16:56:53 2003
@@ -7,7 +7,7 @@
 
 O_TARGET := fs.o
 
-export-objs :=	filesystems.o open.o dcache.o buffer.o
+export-objs :=	filesystems.o open.o dcache.o buffer.o inode.o
 mod-subdirs :=	nls
 
 obj-y :=	open.o read_write.o devices.o file_table.o buffer.o \
--- linux-chaos/mm/filemap.c.b_io_export	Wed Jan 29 16:50:39 2003
+++ linux-chaos/mm/filemap.c	Wed Jan 29 16:51:11 2003
@@ -65,6 +65,7 @@
  *			pagecache_lock
  */
 spinlock_cacheline_t pagemap_lru_lock_cacheline = {SPIN_LOCK_UNLOCKED};
+EXPORT_SYMBOL(pagemap_lru_lock_cacheline);
 
 #define CLUSTER_PAGES		(1 << page_cluster)
 #define CLUSTER_OFFSET(x)	(((x) >> page_cluster) << page_cluster)
--- linux-chaos/mm/vmscan.c.b_io_export	Wed Jan 29 16:51:58 2003
+++ linux-chaos/mm/vmscan.c	Wed Jan 29 16:55:16 2003
@@ -839,6 +839,7 @@
 	set_current_state(TASK_RUNNING);
 	remove_wait_queue(&kswapd_done, &wait);
 }
+EXPORT_SYMBOL(wakeup_kswapd);
 
 static void wakeup_memwaiters(void)
 {
--- linux-chaos/mm/Makefile.b_io_export	Wed Jan 29 16:52:46 2003
+++ linux-chaos/mm/Makefile	Wed Jan 29 16:54:23 2003
@@ -9,7 +9,7 @@
 
 O_TARGET := mm.o
 
-export-objs := shmem.o filemap.o memory.o page_alloc.o mempool.o
+export-objs := shmem.o filemap.o memory.o page_alloc.o mempool.o vmscan.c
 
 obj-y	 := memory.o mmap.o filemap.o mprotect.o mlock.o mremap.o \
 	    vmalloc.o slab.o bootmem.o swap.o vmscan.o page_io.o \
--- linux-chaos/mm/page_alloc.c.b_io_export	Wed Jan 29 17:00:32 2003
+++ linux-chaos/mm/page_alloc.c	Wed Jan 29 17:01:31 2003
@@ -31,6 +31,7 @@
 int nr_inactive_dirty_pages;
 int nr_inactive_clean_pages;
 pg_data_t *pgdat_list;
+EXPORT_SYMBOL(pgdat_list);
 
 /*
  * The zone_table array is used to look up the address of the
