Index: linux-2.6.4-51.0/Documentation/filesystems/ext2.txt
===================================================================
--- linux-2.6.4-51.0.orig/Documentation/filesystems/ext2.txt	2004-05-06 22:21:26.000000000 -0400
+++ linux-2.6.4-51.0/Documentation/filesystems/ext2.txt	2004-05-06 22:24:42.000000000 -0400
@@ -35,6 +35,22 @@
 
 sb=n				Use alternate superblock at this location.
 
+iopen				Makes an invisible pseudo-directory called 
+				__iopen__ available in the root directory
+				of the filesystem.  Allows open-by-inode-
+				number.  i.e., inode 3145 can be accessed
+				via /mntpt/__iopen__/3145
+
+iopen_nopriv			This option makes the iopen directory be
+				world-readable.  This may be safer since it
+				allows daemons to run as an unprivileged user,
+				however it significantly changes the security
+				model of a Unix filesystem, since previously
+				all files under a mode 700 directory were not
+				generally avilable even if the
+				permissions on the file itself is
+				world-readable.
+
 grpquota,noquota,quota,usrquota	Quota options are silently ignored by ext2.
 
 
Index: linux-2.6.4-51.0/fs/dcache.c
===================================================================
--- linux-2.6.4-51.0.orig/fs/dcache.c	2004-05-06 22:24:42.000000000 -0400
+++ linux-2.6.4-51.0/fs/dcache.c	2004-05-06 22:58:37.000000000 -0400
@@ -1195,7 +1195,7 @@
  * dcache entries should not be moved in this way.
  */
 
-static void d_move_locked(struct dentry * dentry, struct dentry * target)
+void d_move_locked(struct dentry * dentry, struct dentry * target)
 {
	struct hlist_head *list;

@@ -1253,6 +1252,8 @@
 	write_sequnlock(&rename_lock);
 }
 
+EXPORT_SYMBOL(d_move_locked);
+
 /**
  * d_move - move a dentry
  * @dentry: entry to move
Index: linux-2.6.4-51.0/include/linux/dcache.h
===================================================================
--- linux-2.6.4-51.0.orig/include/linux/dcache.h	2004-05-06 22:24:42.000000000 -0400
+++ linux-2.6.4-51.0/include/linux/dcache.h	2004-05-06 23:03:43.000000000 -0400
@@ -234,6 +234,7 @@
  * This adds the entry to the hash queues.
  */
 extern void d_rehash(struct dentry *);
+extern void d_rehash_cond(struct dentry *, int lock);
 
 /**
  * d_add - add dentry to hash queues
@@ -252,6 +253,7 @@
 
 /* used for rename() and baskets */
 extern void d_move(struct dentry *, struct dentry *);
+extern void d_move_locked(struct dentry *, struct dentry *);
 
 /* appendix may either be NULL or be used for transname suffixes */
 extern struct dentry * d_lookup(struct dentry *, struct qstr *);
