Index: linux-2.6.22.5/Documentation/filesystems/ext2.txt
===================================================================
--- linux-2.6.22.5.orig/Documentation/filesystems/ext2.txt	2007-08-22 17:23:54.000000000 -0600
+++ linux-2.6.22.5/Documentation/filesystems/ext2.txt	2008-02-21 00:57:02.000000000 -0700
@@ -58,6 +58,22 @@
 
 xip				Use execute in place (no caching) if possible
 
+iopen				Makes an invisible pseudo-directory called 
+				__iopen__ available in the root directory
+				of the filesystem.  Allows open-by-inode-
+				number.  i.e., inode 3145 can be accessed
+				via /mntpt/__iopen__/3145
+
+iopen_nopriv			This option makes the iopen directory be
+				world-readable.  This may be safer since it
+				allows daemons to run as an unprivileged user,
+				however it significantly changes the security
+				model of a Unix filesystem, since previously
+				all files under a mode 700 directory were not
+				generally avilable even if the
+				permissions on the file itself is
+				world-readable.
+
 grpquota,noquota,quota,usrquota	Quota options are silently ignored by ext2.
 
 
Index: linux-2.6.22.5/fs/dcache.c
===================================================================
--- linux-2.6.22.5.orig/fs/dcache.c	2008-02-21 00:56:29.000000000 -0700
+++ linux-2.6.22.5/fs/dcache.c	2008-02-21 00:57:02.000000000 -0700
@@ -1537,7 +1537,7 @@
  * Update the dcache to reflect the move of a file name. Negative
  * dcache entries should not be moved in this way.
  */
-static void d_move_locked(struct dentry * dentry, struct dentry * target)
+void d_move_locked(struct dentry * dentry, struct dentry * target)
 {
 	struct hlist_head *list;
 
@@ -1595,6 +1595,7 @@
 	spin_unlock(&dentry->d_lock);
 	write_sequnlock(&rename_lock);
 }
+EXPORT_SYMBOL(d_move_locked);
 
 /**
  * d_move - move a dentry
Index: linux-2.6.22.5/include/linux/dcache.h
===================================================================
--- linux-2.6.22.5.orig/include/linux/dcache.h	2008-02-21 00:56:29.000000000 -0700
+++ linux-2.6.22.5/include/linux/dcache.h	2008-02-21 00:57:02.000000000 -0700
@@ -251,6 +251,7 @@
  * This adds the entry to the hash queues.
  */
 extern void d_rehash(struct dentry *);
+extern void d_rehash_cond(struct dentry *, int lock);
 
 /**
  * d_add - add dentry to hash queues
@@ -286,6 +287,7 @@
 
 /* used for rename() and baskets */
 extern void d_move(struct dentry *, struct dentry *);
+extern void d_move_locked(struct dentry *, struct dentry *);
 
 /* appendix may either be NULL or be used for transname suffixes */
 extern struct dentry * d_lookup(struct dentry *, struct qstr *);
