
--- linux-2.4.18/fs/jbd/commit.c~jbd-dont-account-blocks-twice	Mon Jul 28 13:52:05 2003
+++ linux-2.4.18-alexey/fs/jbd/commit.c	Mon Jul 28 14:03:53 2003
@@ -407,6 +407,11 @@ sync_datalist_empty:
 			continue;
 		}
 
+		/* start_this_handle() accounts t_outstanding_credits
+		 * to know free space in log, but this counter is changed
+		 * by journal_next_log_block() also. */
+		commit_transaction->t_outstanding_credits--;
+
 		/* Bump b_count to prevent truncate from stumbling over
                    the shadowed buffer!  @@@ This can go if we ever get
                    rid of the BJ_IO/BJ_Shadow pairing of buffers. */

_
