Index: linux/fs/jbd/journal.c
===================================================================
--- linux.orig/fs/jbd/journal.c
+++ linux/fs/jbd/journal.c
@@ -288,7 +288,7 @@ int journal_write_metadata_buffer(transa
 				  struct journal_head **jh_out,
 				  int blocknr)
 {
-	int need_copy_out = 0;
+	int need_copy_out;
 	int done_copy_out = 0;
 	int do_escape = 0;
 	char *mapped_data;
@@ -334,7 +334,8 @@ repeat:
 				cpu_to_be32(JFS_MAGIC_NUMBER)) {
 		need_copy_out = 1;
 		do_escape = 1;
-	}
+	} else
+		need_copy_out = jh_in->b_committed_data == NULL;
 	kunmap_atomic(mapped_data, KM_USER0);
 
 	/*
