Date: Tue, 26 Sep 2006 11:00:28 -0500
From: Eric Sandeen <esandeen@redhat.com>
Subject: Re: [PATCH RHEL5] 16T overflows in jbd code

Signed-off-by: Eric Sandeen <esandeen@redhat.com>
Signed-off-by: Andrew Morton <akpm@osdl.org>

Index: linux-2.6.17-1.2654.el5/fs/jbd/journal.c
===================================================================
--- linux-2.6.17-1.2654.el5.orig/fs/jbd/journal.c
+++ linux-2.6.17-1.2654.el5/fs/jbd/journal.c
@@ -271,7 +271,7 @@ static void journal_kill_thread(journal_
 int journal_write_metadata_buffer(transaction_t *transaction,
 				  struct journal_head  *jh_in,
 				  struct journal_head **jh_out,
-				  int blocknr)
+				  unsigned long blocknr)
 {
 	int need_copy_out = 0;
 	int done_copy_out = 0;
@@ -696,7 +696,7 @@ fail:
  *  @bdev: Block device on which to create the journal
  *  @fs_dev: Device which hold journalled filesystem for this journal.
  *  @start: Block nr Start of journal.
- *  @len:  Lenght of the journal in blocks.
+ *  @len:  Length of the journal in blocks.
  *  @blocksize: blocksize of journalling device
  *  @returns: a newly created journal_t *
  *  
Index: linux-2.6.17-1.2654.el5/include/linux/jbd.h
===================================================================
--- linux-2.6.17-1.2654.el5.orig/include/linux/jbd.h
+++ linux-2.6.17-1.2654.el5/include/linux/jbd.h
@@ -866,7 +866,7 @@ extern int 
 journal_write_metadata_buffer(transaction_t	  *transaction,
 			      struct journal_head  *jh_in,
 			      struct journal_head **jh_out,
-			      int		   blocknr);
+			      unsigned long	   blocknr);
 
 /* Transaction locking */
 extern void		__wait_on_journal (journal_t *);

