Index: linux-2.6.32-71.18.1.el6-master/fs/jbd2/transaction.c
===================================================================
--- linux-2.6.32-71.18.1.el6-master.orig/fs/jbd2/transaction.c	2011-03-18 10:05:45.000000000 +0800
+++ linux-2.6.32-71.18.1.el6-master/fs/jbd2/transaction.c	2011-03-18 10:06:19.001935306 +0800
@@ -59,7 +59,7 @@
 	INIT_LIST_HEAD(&transaction->t_private_list);
 
 	/* Set up the commit timer for the new transaction. */
-	journal->j_commit_timer.expires = round_jiffies_up(transaction->t_expires);
+	journal->j_commit_timer.expires = transaction->t_expires;
 	add_timer(&journal->j_commit_timer);
 
 	J_ASSERT(journal->j_running_transaction == NULL);
