Index: linux-2.4.19-pre1/include/linux/kdev_t.h
===================================================================
--- linux-2.4.19-pre1.orig/include/linux/kdev_t.h	2003-11-20 19:01:38.000000000 +0300
+++ linux-2.4.19-pre1/include/linux/kdev_t.h	2003-11-21 04:02:03.000000000 +0300
@@ -82,6 +82,7 @@
 #define minor(d) MINOR(d)
 #define kdev_same(a,b) (a==b)
 #define kdev_none(d) (!(d))
+#define val_to_kdev(d)  ((kdev_t)(d))
 
 /*
 As long as device numbers in the outside world have 16 bits only,
