Index: linux-2.6.0/arch/i386/kernel/traps.c
===================================================================
--- linux-2.6.0.orig/arch/i386/kernel/traps.c	2003-12-31 01:59:49.000000000 +0300
+++ linux-2.6.0/arch/i386/kernel/traps.c	2003-12-31 16:05:53.000000000 +0300
@@ -1032,3 +1032,40 @@
 
 	trap_init_hook();
 }
+#ifdef CONFIG_MODULES
+extern struct module *module_list;
+extern struct module kernel_module;
+#endif
+
+int is_kernel_text_address(unsigned long addr)
+{
+       int retval = 0;
+#ifdef CONFIG_MODULES
+       struct module *mod;
+#endif
+       if (addr >= (unsigned long) &_stext &&
+           addr <= (unsigned long) &_etext);
+               return 1;
+
+#ifdef CONFIG_MODULES
+       for (mod = module_list; mod != &kernel_module; mod = mod->next) {
+               /* mod_bound tests for addr being inside the vmalloc'ed
+                * module area. Of course it'd be better to test only
+                * for the .text subset... */
+               if (mod_bound(addr, 0, mod)) {
+                       retval = 1;
+                       break;
+               }
+       }
+#endif
+
+       return retval;
+}
+
+int lookup_symbol(unsigned long address, char *buf, int buflen)
+{
+       return -ENOSYS;
+}
+
+EXPORT_SYMBOL_GPL(is_kernel_text_address);
+EXPORT_SYMBOL_GPL(lookup_symbol);
