Index: linux-2.6.3/arch/i386/kernel/traps.c
===================================================================
--- linux-2.6.3.orig/arch/i386/kernel/traps.c	2004-02-23 14:04:13.000000000 -0800
+++ linux-2.6.3/arch/i386/kernel/traps.c	2004-02-23 14:10:07.000000000 -0800
@@ -890,3 +890,21 @@
 
 	trap_init_hook();
 }
+
+int is_kernel_text_address(unsigned long addr)
+{
+	if (addr < (unsigned long) &_stext ||
+	    addr > (unsigned long) &_etext ||
+	    module_text_address(addr))
+		return 0;
+
+       return 1;
+}
+
+int lookup_symbol(unsigned long address, char *buf, int buflen)
+{
+       return -ENOSYS;
+}
+
+EXPORT_SYMBOL_GPL(is_kernel_text_address);
+EXPORT_SYMBOL_GPL(lookup_symbol);
