 arch/i386/Makefile |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

diff -puN arch/i386/Makefile~kgdb-use-ggdb arch/i386/Makefile
--- 25/arch/i386/Makefile~kgdb-use-ggdb	2003-06-14 22:54:41.000000000 -0700
+++ 25-akpm/arch/i386/Makefile	2003-06-14 22:54:41.000000000 -0700
@@ -85,7 +85,7 @@ mcore-$(CONFIG_X86_ES7000)	:= mach-es700
 # default subarch .h files
 mflags-y += -Iinclude/asm-i386/mach-default
 
-mflags-$(CONFIG_KGDB) += -g
+mflags-$(CONFIG_KGDB) += -ggdb
 mflags-$(CONFIG_KGDB_MORE) += $(shell echo $(CONFIG_KGDB_OPTIONS) | sed -e 's/"//g')
 
 head-y := arch/i386/kernel/head.o arch/i386/kernel/init_task.o

_
