 ext2/super.c        |    3 +--
 ext3/ext3-exports.c |   13 +++++++++++++
 2 files changed, 14 insertions(+), 2 deletions(-)

Index: linux-2.4.19.SuSE/fs/ext2/super.c
===================================================================
--- linux-2.4.19.SuSE.orig/fs/ext2/super.c	Mon Jan 27 05:08:00 2003
+++ linux-2.4.19.SuSE/fs/ext2/super.c	Sun Nov 16 00:40:59 2003
@@ -70,6 +70,7 @@
 {
 	va_list args;
 
+	ext2_xattr_put_super(sb);
 	if (!(sb->s_flags & MS_RDONLY)) {
 		sb->u.ext2_sb.s_mount_state |= EXT2_ERROR_FS;
 		sb->u.ext2_sb.s_es->s_state =
Index: linux-2.4.19.SuSE/fs/ext3/super.c
===================================================================
--- linux-2.4.19.SuSE.orig/fs/ext3/super.c	Mon Jan 27 05:08:00 2003
+++ linux-2.4.19.SuSE/fs/ext3/super.c	Sun Nov 16 00:40:59 2003
@@ -1822,8 +1828,6 @@
 	exit_ext3_xattr();
 }
 
-EXPORT_SYMBOL(ext3_force_commit);
-EXPORT_SYMBOL(ext3_bread);
 
 MODULE_AUTHOR("Remy Card, Stephen Tweedie, Andrew Morton, Andreas Dilger, Theodore Ts'o and others");
 MODULE_DESCRIPTION("Second Extended Filesystem with journaling extensions");
Index: linux-2.4.19.SuSE/fs/ext3/ext3-exports.c
===================================================================
--- linux-2.4.19.SuSE.orig/fs/ext3/ext3-exports.c	Sun Nov 16 00:40:58 2003
+++ linux-2.4.19.SuSE/fs/ext3/ext3-exports.c	Sun Nov 16 00:40:59 2003
@@ -0,0 +1,13 @@
+#include <linux/config.h>
+#include <linux/module.h>
+#include <linux/ext3_fs.h>
+#include <linux/ext3_jbd.h>
+#include <linux/ext3_xattr.h>
+
+EXPORT_SYMBOL(ext3_force_commit);
+EXPORT_SYMBOL(ext3_bread);
+EXPORT_SYMBOL(ext3_xattr_register);
+EXPORT_SYMBOL(ext3_xattr_unregister);
+EXPORT_SYMBOL(ext3_xattr_get);
+EXPORT_SYMBOL(ext3_xattr_list);
+EXPORT_SYMBOL(ext3_xattr_set);
