Index: linux-2.4.20/include/linux/list.h
===================================================================
--- linux-2.4.20.orig/include/linux/list.h	2003-10-17 23:56:26.000000000 +0400
+++ linux-2.4.20/include/linux/list.h	2003-10-22 14:02:22.000000000 +0400
@@ -227,6 +227,17 @@
 	     pos = list_entry(pos->member.next, typeof(*pos), member),	\
 		     prefetch(pos->member.next))
 
+/* 2.5 uses hlists for some things, like the d_hash.  we'll treat them
+ * as 2.5 and let macros drop back.. */
+#define hlist_entry                     list_entry
+#define hlist_head                      list_head
+#define hlist_node                      list_head
+#define HLIST_HEAD                      LIST_HEAD
+#define INIT_HLIST_HEAD                 INIT_LIST_HEAD
+#define hlist_del_init                  list_del_init
+#define hlist_add_head                  list_add
+#define hlist_for_each_safe             list_for_each_safe
+
 #endif /* __KERNEL__ || _LVM_H_INCLUDE */
 
 #endif
