Index: linux-2.4.18-chaos/include/linux/list.h
===================================================================
--- linux-2.4.18-chaos.orig/include/linux/list.h	2003-10-22 13:43:39.000000000 +0400
+++ linux-2.4.18-chaos/include/linux/list.h	2003-10-22 13:44:17.000000000 +0400
@@ -173,6 +173,16 @@
 	for (pos = (head)->prev, prefetch(pos->prev); pos != (head); \
         	pos = pos->prev, prefetch(pos->prev))
         	
+/* 2.5 uses hlists for some things, like the d_hash.  we'll treat them
+ * as 2.5 and let macros drop back.. */
+#define hlist_entry                     list_entry
+#define hlist_head                      list_head
+#define hlist_node                      list_head
+#define HLIST_HEAD                      LIST_HEAD
+#define INIT_HLIST_HEAD                 INIT_LIST_HEAD
+#define hlist_del_init                  list_del_init
+#define hlist_add_head                  list_add
+#define hlist_for_each_safe             list_for_each_safe
 
 #endif /* __KERNEL__ || _LVM_H_INCLUDE */
 
