diff -rupN linux-2.6.6.old/fs/block_dev.c linux-2.6.6.new/fs/block_dev.c
--- linux-2.6.6.old/fs/block_dev.c	Mon May 10 05:33:22 2004
+++ linux-2.6.6.new/fs/block_dev.c	Thu Jun 24 20:34:45 2004
@@ -834,6 +834,7 @@ struct block_device *lookup_bdev(const c
 	if (!path || !*path)
 		return ERR_PTR(-EINVAL);
 
+	intent_init(&nd.intent, IT_LOOKUP);
 	error = path_lookup(path, LOOKUP_FOLLOW, &nd);
 	if (error)
 		return ERR_PTR(error);
