Index: linux-2.6.7/fs/block_dev.c
===================================================================
--- linux-2.6.7.orig/fs/block_dev.c	2004-06-16 13:20:26.000000000 +0800
+++ linux-2.6.7/fs/block_dev.c	2004-08-30 17:36:57.000000000 +0800
@@ -832,6 +832,7 @@
 	if (!path || !*path)
 		return ERR_PTR(-EINVAL);
 
+	intent_init(&nd.intent.open, IT_LOOKUP);
 	error = path_lookup(path, LOOKUP_FOLLOW, &nd);
 	if (error)
 		return ERR_PTR(error);
