 fs/Kconfig  |    8 ++++++++
 fs/Makefile |    1 +
 2 files changed, 9 insertions(+)

--- linux-2.5.72/fs/Makefile~lustre_build	2003-06-16 22:20:05.000000000 -0600
+++ linux-2.5.72-braam/fs/Makefile	2003-06-22 10:11:57.000000000 -0600
@@ -57,6 +57,7 @@ obj-$(CONFIG_RAMFS)		+= ramfs/
 obj-$(CONFIG_HUGETLBFS)		+= hugetlbfs/
 obj-$(CONFIG_CODA_FS)		+= coda/
 obj-$(CONFIG_INTERMEZZO_FS)	+= intermezzo/
+obj-$(CONFIG_LUSTRE_FS)		+= lustre/
 obj-$(CONFIG_MINIX_FS)		+= minix/
 obj-$(CONFIG_FAT_FS)		+= fat/
 obj-$(CONFIG_UMSDOS_FS)		+= umsdos/
--- linux-2.5.72/fs/Kconfig~lustre_build	2003-06-16 22:20:05.000000000 -0600
+++ linux-2.5.72-braam/fs/Kconfig	2003-06-22 10:47:15.000000000 -0600
@@ -1561,6 +1561,14 @@ config CODA_FS
 	  whenever you want), say M here and read
 	  <file:Documentation/modules.txt>.  The module will be called coda.
 
+config LUSTRE_FS
+	bool "Lustre: next generation clustering file system (EXPERIMENTAL)"
+	depends on INET && EXPERIMENTAL
+	help
+	  Lustre is a next generation storage architecture which includes a
+	  POSIX compliant cluster file system. For details see
+	  <http://www.lustre.org/>.
+
 config INTERMEZZO_FS
 	tristate "InterMezzo file system support (replicating fs) (EXPERIMENTAL)"
 	depends on INET && EXPERIMENTAL

_
