Index: linux-2.6.27.21-0.1/drivers/md/raid5.c
===================================================================
--- linux-2.6.27.21-0.1.orig/drivers/md/raid5.c	2009-04-23 02:12:52.000000000 -0600
+++ linux-2.6.27.21-0.1/drivers/md/raid5.c	2009-05-22 08:38:38.000000000 -0600
@@ -1760,6 +1760,8 @@
 		bi->bi_next = *bip;
 	*bip = bi;
 	bi->bi_phys_segments++;
+	if (bio_rw_flagged(bi, BIO_RW_SYNCIO) && !forwrite)
+		clear_bit(R5_UPTODATE, &sh->dev[dd_idx].flags); /* force to read from disk. */
 	spin_unlock_irq(&conf->device_lock);
 	spin_unlock(&sh->lock);
 
@@ -3513,6 +3515,8 @@
 
 		bio_endio(bi, 0);
 	}
+	if (bio_rw_flagged(bi, BIO_RW_SYNCIO))
+		raid5_unplug_device(q);
 	return 0;
 }
 
