Index: linux-2.6.9-78.0.22/drivers/md/raid5.c
===================================================================
--- linux-2.6.9-78.0.22.orig/drivers/md/raid5.c	2009-05-11 23:23:39.000000000 -0600
+++ linux-2.6.9-78.0.22/drivers/md/raid5.c	2009-05-11 23:25:08.000000000 -0600
@@ -928,6 +928,8 @@
 #endif
 	*bip = bi;
 	bi->bi_phys_segments ++;
+	if (bio_sync(bi) && !forwrite)
+		clear_bit(R5_UPTODATE, &sh->dev[dd_idx].flags); /* force to read from disk. */
 	spin_unlock_irq(&conf->device_lock);
 	spin_unlock(&sh->lock);
 
@@ -1676,6 +1678,8 @@
 		bi->bi_end_io(bi, bytes, 0);
 	}
 	spin_unlock_irq(&conf->device_lock);
+	if (bio_sync(bi))
+		raid5_unplug_device(q);
 	return 0;
 }
 
