Index: linux-2.4.20-30.9/scripts/mkdep.c
===================================================================
--- linux-2.4.20-30.9.orig/scripts/mkdep.c	2004-02-19 19:40:51.000000000 -0500
+++ linux-2.4.20-30.9/scripts/mkdep.c	2004-04-28 17:24:54.000000000 -0400
@@ -48,8 +48,6 @@
 char __depname[512] = "\n\t@touch ";
 #define depname (__depname+9)
 int hasdep;
-char cwd[PATH_MAX];
-int lcwd;
 
 struct path_struct {
 	int len;
@@ -204,22 +202,8 @@
 		memcpy(path->buffer+path->len, name, len);
 		path->buffer[path->len+len] = '\0';
 		if (access(path->buffer, F_OK) == 0) {
-			int l = lcwd + strlen(path->buffer);
-			char name2[l+2], *p;
-			if (path->buffer[0] == '/') {
-				memcpy(name2, path->buffer, l+1);
-			}
-			else {
-				memcpy(name2, cwd, lcwd);
-				name2[lcwd] = '/';
-				memcpy(name2+lcwd+1, path->buffer, path->len+len+1);
-			}
-			while ((p = strstr(name2, "/../"))) {
-				*p = '\0';
-				strcpy(strrchr(name2, '/'), p+3);
-			}
 			do_depname();
-			printf(" \\\n   %s", name2);
+			printf(" \\\n   %s", path->buffer);
 			return;
 		}
 	}
@@ -601,12 +585,6 @@
 		return 1;
 	}
 
-	if (!getcwd(cwd, sizeof(cwd))) {
-		fprintf(stderr, "mkdep: getcwd() failed %m\n");
-		return 1;
-	}
-	lcwd = strlen(cwd);
-
 	add_path(".");		/* for #include "..." */
 
 	while (++argv, --argc > 0) {
