Increase MAX_SGE for fusion mpt driver.

Index: linux-2.6.32.i386/drivers/message/fusion/Kconfig
===================================================================
--- linux-2.6.32.i386.orig/drivers/message/fusion/Kconfig	2009-12-03 09:21:21.000000000 +0530
+++ linux-2.6.32.i386/drivers/message/fusion/Kconfig	2010-03-16 16:45:08.000000000 +0530
@@ -61,9 +61,9 @@
 	  LSISAS1078
 
 config FUSION_MAX_SGE
-	int "Maximum number of scatter gather entries (16 - 128)"
-	default "128"
-	range 16 128
+	int "Maximum number of scatter gather entries (16 - 256)"
+	default "256"
+	range 16 256
 	help
 	  This option allows you to specify the maximum number of scatter-
 	  gather entries per I/O. The driver default is 128, which matches
Index: linux-2.6.32.i386/drivers/message/fusion/mptbase.h
===================================================================
--- linux-2.6.32.i386.orig/drivers/message/fusion/mptbase.h	2009-12-03 09:21:21.000000000 +0530
+++ linux-2.6.32.i386/drivers/message/fusion/mptbase.h	2010-03-16 16:46:54.000000000 +0530
@@ -165,10 +165,10 @@
  * Set the MAX_SGE value based on user input.
  */
 #ifdef CONFIG_FUSION_MAX_SGE
-#if CONFIG_FUSION_MAX_SGE  < 16
+#if CONFIG_FUSION_MAX_SGE < 16
 #define MPT_SCSI_SG_DEPTH	16
-#elif CONFIG_FUSION_MAX_SGE  > 128
-#define MPT_SCSI_SG_DEPTH	128
+#elif CONFIG_FUSION_MAX_SGE > 256
+#define MPT_SCSI_SG_DEPTH	256
 #else
 #define MPT_SCSI_SG_DEPTH	CONFIG_FUSION_MAX_SGE
 #endif
