 0 files changed

--- linux-2.4.20-rh/drivers/net/netconsole.c~netconsole_sysrq	2003-04-11 14:04:57.000000000 +0800
+++ linux-2.4.20-rh-root/drivers/net/netconsole.c	2003-07-01 11:10:26.000000000 +0800
@@ -988,7 +988,15 @@ static void netconsole_netdump (struct p
 			reply.info = 0;
 			send_netdump_skb(dev, tmp, strlen(tmp), &reply);
 			break;
-
+		 case COMM_SYSRQ:
+			Dprintk("got SYSRQ command.\n");
+			printk("netdump: got SYSRQ command %d \n", req->from);
+			handle_sysrq(req->from, regs, NULL, NULL);
+			reply.code = REPLY_SYSRQ;
+			reply.nr = req->nr;
+			reply.info = req->from;
+			send_netdump_skb(dev, tmp, strlen(tmp), &reply);
+			break;
 		default:
 			reply.code = REPLY_ERROR;
 			reply.nr = req->nr;
--- linux-2.4.20-rh/drivers/net/netconsole.h~netconsole_sysrq	2003-04-11 14:04:57.000000000 +0800
+++ linux-2.4.20-rh-root/drivers/net/netconsole.h	2003-07-01 11:11:29.000000000 +0800
@@ -42,6 +42,7 @@ enum netdump_commands {
 	COMM_START_NETDUMP_ACK = 7,
 	COMM_GET_REGS = 8,
 	COMM_SHOW_STATE = 9,
+	COMM_SYSRQ=10,
 };
 
 #define NETDUMP_REQ_SIZE (8+4*4)
@@ -69,6 +70,7 @@ enum netdump_replies {
 	REPLY_REGS = 10,
 	REPLY_MAGIC = 11,
 	REPLY_SHOW_STATE = 12,
+	REPLY_SYSRQ=13,
 };
 
 typedef struct netdump_reply_s {

_
