Index: linux-2.4.21/fs/nfs/inode.c
===================================================================
--- linux-2.4.21.orig/fs/nfs/inode.c	2005-06-01 22:51:55.000000000 -0400
+++ linux-2.4.21/fs/nfs/inode.c	2005-06-01 23:38:54.883239912 -0400
@@ -679,14 +679,21 @@
 		goto too_big;
 	if (TOOBIG(((res.abytes + blockres) >> blockbits)))
 		goto too_big;
-	if (TOOBIG(res.tfiles) || TOOBIG(res.afiles))
-		goto too_big;
 
 	buf->f_blocks = (res.tbytes + blockres) >> blockbits;
 	buf->f_bfree = (res.fbytes + blockres) >> blockbits;
 	buf->f_bavail = (res.abytes + blockres) >> blockbits;
-	buf->f_files = res.tfiles;
-	buf->f_ffree = res.afiles;
+
+	if (TOOBIG(res.tfiles))
+		buf->f_files = -1;
+	else
+		buf->f_files = res.tfiles;
+
+	if (TOOBIG(res.afiles))
+		buf->f_ffree = -1;
+	else
+		buf->f_ffree = res.afiles;
+
 	return 0;
 
  too_big:
