===== fs/nfsd/vfs.c 1.20 vs edited =====
--- 1.20/fs/nfsd/vfs.c	2005-02-08 16:35:28 +02:00
+++ edited/fs/nfsd/vfs.c	2005-05-29 00:46:44 +03:00
@@ -297,6 +297,7 @@
 	iap->ia_valid |= ATTR_CTIME;
 
 	if (iap->ia_valid & ATTR_SIZE) {
+		down_write(&inode->i_alloc_sem);
 		fh_lock(fhp);
 		size_change = 1;
 	}
@@ -307,6 +308,7 @@
 	}
 	if (size_change) {
 		fh_unlock(fhp);
+		up_write(&inode->i_alloc_sem);
 		put_write_access(inode);
 	}
 	if (!err)
