Index: linux-2.6.10/kernel/sched.c
===================================================================
--- linux-2.6.10.orig/include/linux/sched.h	2005-04-06 15:38:35.000000000 +0000
+++ linux-2.6.10/include/linux/sched.h	2005-07-06 15:06:06.000000000 +0000
@@ -365,6 +376,7 @@ struct user_struct {
 	/* Hash table maintenance information */
 	struct list_head uidhash_list;
 	uid_t uid;
+       __u64 nid;
 };
 
 extern struct user_struct *find_user(uid_t);
