Index: linux-2.4.21-chaos/include/linux/swap.h
===================================================================
--- linux-2.4.21-chaos.orig/include/linux/swap.h	2003-12-12 16:24:33.000000000 +0300
+++ linux-2.4.21-chaos/include/linux/swap.h	2003-12-12 16:41:15.000000000 +0300
@@ -95,7 +95,7 @@
 extern atomic_t page_cache_size;
 extern atomic_t buffermem_pages;
 
-#if 1
+#if 0
 
 static inline void
 	lock_pagecache(void) { br_write_lock(BR_PAGECACHE_LOCK); }
@@ -110,6 +110,7 @@
 
 extern spinlock_cacheline_t pagecache_lock_cacheline;
 #define __pagecache_lock (pagecache_lock_cacheline.lock)
+#define pagecache_lock (pagecache_lock_cacheline.lock)
 
 static inline void
 	lock_pagecache(void) { spin_lock(&__pagecache_lock); }
