Index: linux-2.4.18-chaos/include/linux/proc_fs.h
===================================================================
--- linux-2.4.18-chaos.orig/include/linux/proc_fs.h	2003-07-28 17:52:18.000000000 +0400
+++ linux-2.4.18-chaos/include/linux/proc_fs.h	2004-01-12 20:12:38.000000000 +0300
@@ -25,7 +25,7 @@
 /* Finally, the dynamically allocatable proc entries are reserved: */
 
 #define PROC_DYNAMIC_FIRST 4096
-#define PROC_NDYNAMIC      4096
+#define PROC_NDYNAMIC      16384 
 
 #define PROC_SUPER_MAGIC 0x9fa0
 
