Index: linux-2.4.21-suse2/include/linux/proc_fs.h
===================================================================
--- linux-2.4.21-suse2.orig/include/linux/proc_fs.h	2004-01-12 09:56:51.000000000 +0300
+++ linux-2.4.21-suse2/include/linux/proc_fs.h	2004-01-12 20:25:55.000000000 +0300
@@ -25,11 +25,7 @@
 /* Finally, the dynamically allocatable proc entries are reserved: */
 
 #define PROC_DYNAMIC_FIRST 4096
-#ifdef CONFIG_PPC64
 #define PROC_NDYNAMIC      16384
-#else
-#define PROC_NDYNAMIC      4096
-#endif
 
 #define PROC_SUPER_MAGIC 0x9fa0
 
