--- linux-2.6.9.orig/drivers/md/raid6main.c	2006-09-07 23:12:09.000000000 +0800
+++ linux-2.6.9/drivers/md/raid6main.c	2006-09-07 23:12:44.000000000 +0800
@@ -1775,6 +1775,11 @@ static int run (mddev_t *mddev)
 	mddev->queue->unplug_fn = raid6_unplug_device;
 	mddev->queue->issue_flush_fn = raid6_issue_flush;
 
+	/* in order to support large I/Os */
+	blk_queue_max_sectors(mddev->queue, mddev->chunk_size * mddev->raid_disks >> 9);
+	mddev->queue->max_phys_segments = mddev->chunk_size * mddev->raid_disks >> PAGE_SHIFT;
+	mddev->queue->max_hw_segments = mddev->chunk_size * mddev->raid_disks >> PAGE_SHIFT;
+
 	PRINTK("raid6: run(%s) called.\n", mdname(mddev));
 
 	ITERATE_RDEV(mddev,rdev,tmp) {
