Index: linux-2.4.19-pre1/include/linux/sched.h
===================================================================
--- linux-2.4.19-pre1.orig/include/linux/sched.h	2003-11-21 04:05:05.000000000 +0300
+++ linux-2.4.19-pre1/include/linux/sched.h	2003-11-21 04:10:29.000000000 +0300
@@ -927,6 +927,11 @@
 	return res;
 }
 
+static inline int need_resched(void)
+{
+	        return (unlikely(current->need_resched));
+}
+
 #endif /* __KERNEL__ */
 
 #endif
