Index: linux-2.6.10/include/scsi/scsi.h
===================================================================
--- linux-2.6.10.orig/include/scsi/scsi.h	2004-12-25 00:35:40.000000000 +0300
+++ linux-2.6.10/include/scsi/scsi.h	2005-09-05 18:05:35.000000000 +0400
@@ -14,7 +14,11 @@
  *	The maximum sg list length SCSI can cope with
  *	(currently must be a power of 2 between 32 and 256)
  */
-#define SCSI_MAX_PHYS_SEGMENTS	MAX_PHYS_SEGMENTS
+
+/*
+ * bump it to 256 (was MAX_PHYS_SEGMENTS) for 1MB bio's -bzzz
+ */
+#define SCSI_MAX_PHYS_SEGMENTS	256
 
 
 /*
