Index: linux-2.4.21/fs/open.c
===================================================================
--- linux-2.4.21.orig/fs/open.c	2005-05-04 16:09:11.702098704 -0400
+++ linux-2.4.21/fs/open.c	2005-05-04 16:17:31.597103112 -0400
@@ -92,15 +92,15 @@
 		if (retval)
 			return retval;
 		/* Stuff the 32 bit values into the 64 bit struct */
-		buf->f_type = st.f_type;
-		buf->f_bsize = st.f_bsize;
-		buf->f_blocks = st.f_blocks;
-		buf->f_bfree = st.f_bfree;
-		buf->f_bavail = st.f_bavail;
-		buf->f_files = st.f_files;
-		buf->f_ffree = st.f_ffree;
+		buf->f_type = (unsigned long) st.f_type;
+		buf->f_bsize = (unsigned long) st.f_bsize;
+		buf->f_blocks = (unsigned long) st.f_blocks;
+		buf->f_bfree = (unsigned long) st.f_bfree;
+		buf->f_bavail = (unsigned long) st.f_bavail;
+		buf->f_files = (unsigned long) st.f_files;
+		buf->f_ffree = (unsigned long) st.f_ffree;
 		buf->f_fsid = st.f_fsid;
-		buf->f_namelen = st.f_namelen;
+		buf->f_namelen = (unsigned long) st.f_namelen;
 		memset(buf->f_spare, 0, sizeof(buf->f_spare));
 	}
 	return 0;
