Index: linux-2.4.24/arch/um/os-Linux/process.c
===================================================================
--- linux-2.4.24.orig/arch/um/os-Linux/process.c	2005-02-14 14:25:32.784128506 -0500
+++ linux-2.4.24/arch/um/os-Linux/process.c	2005-02-14 14:26:24.990396165 -0500
@@ -7,6 +7,7 @@
 #include <stdio.h>
 #include <errno.h>
 #include <signal.h>
+#include <linux/unistd.h>
 #include <sys/mman.h>
 #include <sys/wait.h>
 #include "os.h"
@@ -92,7 +93,8 @@
 
 void os_usr1_process(int pid)
 {
-	kill(pid, SIGUSR1);
+       syscall(__NR_tkill, pid, SIGUSR1);
+/*     tkill(pid, SIGUSR1);*/
 }
 
 int os_getpid(void)
