 arch/um/kernel/process.c |    2 +-
 1 files changed, 1 insertion(+), 1 deletion(-)

--- linux-2.4.20/arch/um/kernel/process.c~uml-patch-2.4.20-tmp	2003-04-15 11:13:26.000000000 +0800
+++ linux-2.4.20-root/arch/um/kernel/process.c	2003-04-15 11:24:21.000000000 +0800
@@ -72,7 +72,7 @@ void init_new_thread_signals(int altstac
 		    SIGUSR1, SIGIO, SIGWINCH, SIGALRM, SIGVTALRM, -1);
 	set_handler(SIGUSR2, (__sighandler_t) sig_handler, 
 		    SA_NOMASK | flags, -1);
-	(void) CHOOSE_MODE(signal(SIGCHLD, SIG_IGN), (void *) 0);
+//	(void) CHOOSE_MODE(signal(SIGCHLD, SIG_IGN), (void *) 0);
 	signal(SIGHUP, SIG_IGN);
 
 	init_irq_signals(altstack);

_
