Index: linux-2.6.0/include/asm-um/unistd.h
===================================================================
--- linux-2.6.0.orig/include/asm-um/unistd.h	2004-02-03 18:07:34.000000000 +0300
+++ linux-2.6.0/include/asm-um/unistd.h	2004-02-03 20:29:49.000000000 +0300
@@ -6,8 +6,10 @@
 #ifndef _UM_UNISTD_H_
 #define _UM_UNISTD_H_
 
+#ifdef __KERNEL__
 #include "linux/resource.h"
 #include "asm/uaccess.h"
+#endif
 
 extern long sys_open(const char *filename, int flags, int mode);
 extern long sys_dup(unsigned int fildes);
