Index: linux-2.6.3/arch/i386/kernel/sys_i386.c
===================================================================
--- linux-2.6.3.orig/arch/i386/kernel/sys_i386.c	2004-02-23 14:21:03.000000000 -0800
+++ linux-2.6.3/arch/i386/kernel/sys_i386.c	2004-02-23 14:24:38.000000000 -0800
@@ -56,7 +56,7 @@
 	}
 
 	down_write(&current->mm->mmap_sem);
-	error = do_mmap_pgoff(file, addr, len, prot, flags, pgoff);
+	error = do_mmap_pgoff(current->mm, file, addr, len, prot, flags, pgoff);
 	up_write(&current->mm->mmap_sem);
 
 	if (file)
