Index: linux-2.4.24-vanilla/arch/um/kernel/ksyms.c
===================================================================
--- linux-2.4.24-vanilla.orig/arch/um/kernel/ksyms.c	2004-01-10 17:47:10.000000000 +0300
+++ linux-2.4.24-vanilla/arch/um/kernel/ksyms.c	2004-01-10 18:22:30.000000000 +0300
@@ -34,6 +34,7 @@
 EXPORT_SYMBOL(host_task_size);
 EXPORT_SYMBOL(arch_validate);
 EXPORT_SYMBOL(get_kmem_end);
+EXPORT_SYMBOL(end_iomem);
 
 EXPORT_SYMBOL(high_physmem);
 EXPORT_SYMBOL(empty_zero_page);
