 include/asm-um/unistd.h |    2 ++
 1 files changed, 2 insertions(+)

diff -puN include/asm-um/unistd.h~uml-2.6.0-fix include/asm-um/unistd.h
--- linux-2.6.0-test3/include/asm-um/unistd.h~uml-2.6.0-fix	2003-09-04 18:39:45.000000000 +0400
+++ linux-2.6.0-test3-alexey/include/asm-um/unistd.h	2003-09-04 18:39:59.000000000 +0400
@@ -6,8 +6,10 @@
 #ifndef _UM_UNISTD_H_
 #define _UM_UNISTD_H_
 
+#ifdef __KERNEL__
 #include "linux/resource.h"
 #include "asm/uaccess.h"
+#endif
 
 extern long sys_open(const char *filename, int flags, int mode);
 extern long sys_dup(unsigned int fildes);

_
