--- uml-2.4.24/arch/um/kernel/tt/ksyms.c.orig	2005-05-04 13:59:58.806659456 +0300
+++ uml-2.4.24/arch/um/kernel/tt/ksyms.c	2005-05-04 14:00:18.358687096 +0300
@@ -12,6 +12,8 @@
 EXPORT_SYMBOL(__do_strncpy_from_user);
 EXPORT_SYMBOL(__do_strnlen_user); 
 EXPORT_SYMBOL(__do_clear_user);
+EXPORT_SYMBOL(clear_user_tt);
+EXPORT_SYMBOL(clear_user_skas);
 
 EXPORT_SYMBOL(tracing_pid);
 EXPORT_SYMBOL(honeypot);
