


 arch/um/kernel/mem.c |   15 +++++++++++++++
 1 files changed, 15 insertions(+)

--- linux-2.4.20/arch/um/kernel/mem.c~uml_check_get_page	2003-04-08 23:34:50.000000000 -0600
+++ linux-2.4.20-braam/arch/um/kernel/mem.c	2003-04-08 23:34:50.000000000 -0600
@@ -712,6 +712,21 @@ struct page *pte_mem_map(pte_t pte)
 	return(phys_mem_map(pte_val(pte)));
 }
 
+struct page *check_get_page(unsigned long kaddr)
+{
+	struct page *page;
+	struct mem_region *mr;
+	unsigned long phys = __pa(kaddr);
+	unsigned int n = phys_region_index(phys);
+
+	if (regions[n] == NULL)
+                return NULL;
+
+	mr = regions[n];
+	page = (struct page *) mr->mem_map;
+	return page + ((phys_addr(phys)) >> PAGE_SHIFT);
+}
+
 struct mem_region *page_region(struct page *page, int *index_out)
 {
 	int i;

_
