Index: linux-2.6.10/fs/namespace.c
===================================================================
--- linux-2.6.10.orig/fs/namespace.c	2005-03-31 17:03:37.000000000 +0800
+++ linux-2.6.10/fs/namespace.c	2005-03-31 17:58:42.827926064 +0800
@@ -365,7 +365,7 @@
 	}
 }
 
-static int do_umount(struct vfsmount *mnt, int flags)
+int do_umount(struct vfsmount *mnt, int flags)
 {
 	struct super_block * sb = mnt->mnt_sb;
 	int retval;
@@ -458,6 +458,8 @@
 	return retval;
 }
 
+EXPORT_SYMBOL(do_umount);
+
 /*
  * Now umount can handle mount points as well as block devices.
  * This is important for filesystems which use unnamed block devices.
Index: linux-2.6.10/include/linux/mount.h
===================================================================
--- linux-2.6.10.orig/include/linux/mount.h	2005-03-31 17:15:40.000000000 +0800
+++ linux-2.6.10/include/linux/mount.h	2005-03-31 17:59:41.914943472 +0800
@@ -70,6 +70,7 @@
 extern struct vfsmount *do_kern_mount(const char *fstype, int flags,
 				      const char *name, void *data);
 
+extern int do_umount(struct vfsmount *mnt, int flags);
 struct nameidata;
 
 extern int do_add_mount(struct vfsmount *newmnt, struct nameidata *nd,
