Index: uml-2.6.3/include/linux/dcache.h
===================================================================
--- uml-2.6.3.orig/include/linux/dcache.h	2004-02-21 00:19:14.365988600 +0800
+++ uml-2.6.3/include/linux/dcache.h	2004-02-21 00:21:04.612228624 +0800
@@ -25,6 +25,11 @@
 
 #define IS_ROOT(x) ((x) == (x)->d_parent)
 
+struct dentry_params {
+	unsigned long   p_inum;
+	void            *p_ptr;
+};
+
 /*
  * "quick string" -- eases parameter passing, but more importantly
  * saves "metadata" about the string (ie length and the hash).
