Index: linux-2.6.10/include/linux/dcache.h
===================================================================
--- linux-2.6.10.orig/include/linux/dcache.h	2005-07-11 17:33:07.000000000 +0800
+++ linux-2.6.10/include/linux/dcache.h	2005-07-14 20:27:46.000000000 +0800
@@ -24,6 +24,13 @@
 
 #define IS_ROOT(x) ((x) == (x)->d_parent)
 
+struct dentry_params {
+	unsigned long   p_inum;
+	void            *p_ptr;
+	__u64		p_fid;
+	__u64		p_group;
+};
+
 /*
  * "quick string" -- eases parameter passing, but more importantly
  * saves "metadata" about the string (ie length and the hash).
