Index: linux-2.6.7/fs/namespace.c
===================================================================
--- linux-2.6.7.orig/fs/namespace.c	2004-11-21 00:25:13.000000000 +0200
+++ linux-2.6.7/fs/namespace.c	2004-11-21 00:25:15.000000000 +0200
@@ -360,7 +360,7 @@
 	}
 }
 
-static int do_umount(struct vfsmount *mnt, int flags)
+int do_umount(struct vfsmount *mnt, int flags)
 {
 	struct super_block * sb = mnt->mnt_sb;
 	int retval;
@@ -434,6 +434,8 @@
 	return retval;
 }
 
+EXPORT_SYMBOL(do_umount);
+
 /*
  * Now umount can handle mount points as well as block devices.
  * This is important for filesystems which use unnamed block devices.
Index: linux-2.6.7/include/linux/mount.h
===================================================================
--- linux-2.6.7.orig/include/linux/mount.h	2004-11-21 00:25:13.000000000 +0200
+++ linux-2.6.7/include/linux/mount.h	2005-01-11 15:28:26.627030408 +0200
@@ -56,6 +56,7 @@
 extern struct vfsmount *alloc_vfsmnt(const char *name);
 extern struct vfsmount *do_kern_mount(const char *fstype, int flags,
 				      const char *name, void *data);
+extern int do_umount(struct vfsmount *mnt, int flags);
 extern spinlock_t vfsmount_lock;
 
 #endif
