diff -urp linux-2.6.18.rawops/fs/dcache.c linux-2.6.18.races/fs/dcache.c
--- linux-2.6.18.rawops/fs/dcache.c	2007-02-08 19:00:31.000000000 +0200
+++ linux-2.6.18.races/fs/dcache.c	2007-02-14 19:23:49.000000000 +0200
@@ -230,6 +230,13 @@ int d_invalidate(struct dentry * dentry)
 		spin_unlock(&dcache_lock);
 		return 0;
 	}
+
+	/* network invalidation by Lustre */
+	if (dentry->d_flags & DCACHE_LUSTRE_INVALID) {
+		spin_unlock(&dcache_lock);
+		return 0;
+	}
+
 	/*
 	 * Check whether to do a partial shrink_dcache
 	 * to get rid of unused child entries.
@@ -1400,13 +1407,21 @@ static void _d_rehash(struct dentry * en
  * Adds a dentry to the hash according to its name.
  */
  
-void d_rehash(struct dentry * entry)
+void d_rehash_cond(struct dentry * entry, int lock)
 {
-	spin_lock(&dcache_lock);
+ 	if (lock)
+	    spin_lock(&dcache_lock);
 	spin_lock(&entry->d_lock);
 	_d_rehash(entry);
 	spin_unlock(&entry->d_lock);
-	spin_unlock(&dcache_lock);
+ 	if (lock)	
+	    spin_unlock(&dcache_lock);
+}
+EXPORT_SYMBOL(d_rehash_cond);
+
+void d_rehash(struct dentry * entry)
+{
+	d_rehash_cond(entry, 1);
 }
 
 #define do_switch(x,y) do { \
diff -urp linux-2.6.18.rawops/include/linux/dcache.h linux-2.6.18.races/include/linux/dcache.h
--- linux-2.6.18.rawops/include/linux/dcache.h	2007-02-14 16:52:37.000000000 +0200
+++ linux-2.6.18.races/include/linux/dcache.h	2007-02-14 19:21:14.000000000 +0200
@@ -177,6 +177,7 @@ d_iput:		no		no		no       yes
 
 #define DCACHE_REFERENCED	0x0008  /* Recently used, don't discard. */
 #define DCACHE_UNHASHED		0x0010	
+#define DCACHE_LUSTRE_INVALID	0x0040  /* Lustre invalidated */
 
 #define DCACHE_INOTIFY_PARENT_WATCHED	0x0020 /* Parent inode is watched */
 
@@ -254,6 +255,7 @@ extern int have_submounts(struct dentry 
  * This adds the entry to the hash queues.
  */
 extern void d_rehash(struct dentry *);
+extern void d_rehash_cond(struct dentry *, int lock);
 
 /**
  * d_add - add dentry to hash queues
