Index: linux-2.6.15.i686-cy/include/linux/dcache.h
===================================================================
--- linux-2.6.15.i686-cy.orig/include/linux/dcache.h	2006-06-01 23:07:05.000000000 +0800
+++ linux-2.6.15.i686-cy/include/linux/dcache.h	2006-06-01 23:10:21.000000000 +0800
@@ -4,7 +4,6 @@
 #ifdef __KERNEL__
 
 #include <asm/atomic.h>
-#include <linux/string.h>
 #include <linux/list.h>
 #include <linux/spinlock.h>
 #include <linux/cache.h>
@@ -42,6 +41,7 @@
 	const unsigned char *name;
 };
 
+struct inode;
 #include <linux/namei.h>
 
 struct dentry_stat_t {
