Index: linux-4.18.0-553.16.1.el8_10/drivers/scsi/virtio_scsi.c
===================================================================
--- linux-4.18.0-553.16.1.el8_10.orig/drivers/scsi/virtio_scsi.c
+++ linux-4.18.0-553.16.1.el8_10/drivers/scsi/virtio_scsi.c
@@ -455,7 +455,8 @@ static int __virtscsi_add_cmd(struct vir
 		sgs[out_num + in_num++] = in->sgl;
 	}
 
-	return virtqueue_add_sgs(vq, sgs, out_num, in_num, cmd, GFP_ATOMIC);
+	return virtqueue_add_sgs(vq, sgs, out_num, in_num, cmd,
+				 GFP_ATOMIC | __GFP_NOWARN);
 }
 
 static void virtscsi_kick_vq(struct virtio_scsi_vq *vq)
