Index: linux+rhel4+chaos/mm/page_alloc.c
===================================================================
--- linux+rhel4+chaos.orig/mm/page_alloc.c
+++ linux+rhel4+chaos/mm/page_alloc.c
@@ -1972,8 +1972,12 @@ static void setup_per_zone_pages_min(voi
 			                   lowmem_pages;
 		}
 
-		zone->pages_low = zone->pages_min * 2;
-		zone->pages_high = zone->pages_min * 3;
+		/*
+		 * When interpreting these watermarks, just keep in mind that:
+		 * zone->pages_min == (zone->pages_min * 4) / 4;
+		 */
+		zone->pages_low  = (zone->pages_min * 5) / 4;
+		zone->pages_high = (zone->pages_min * 6) / 4;
 		spin_unlock_irqrestore(&zone->lru_lock, flags);
 	}
 }
