Index: linux.mcp2/kernel/ksyms.c
===================================================================
--- linux.mcp2.orig/kernel/ksyms.c	2004-05-05 14:57:48.000000000 -0700
+++ linux.mcp2/kernel/ksyms.c	2004-05-05 15:32:44.000000000 -0700
@@ -108,6 +108,7 @@
 EXPORT_SYMBOL(kfree);
 EXPORT_SYMBOL(vfree);
 EXPORT_SYMBOL(__vmalloc);
+extern struct page * vmalloc_to_page(void *addr);
 EXPORT_SYMBOL(vmalloc_to_page);
 EXPORT_SYMBOL(mem_map);
 EXPORT_SYMBOL(remap_page_range);
