--- l1/fs/ext4/inode.c	2016-08-23 23:38:23.238091149 +0800
+++ l2/fs/ext4/inode.c	2017-01-24 04:04:45.100092653 +0800
@@ -2683,6 +2683,9 @@ static int ext4_writepages(struct addres
 		mpd.last_page = wbc->range_end >> PAGE_CACHE_SHIFT;
 	}
 
+	ret = ext4_inode_attach_jinode(inode);
+	if (ret)
+		goto out_writepages;
 	mpd.inode = inode;
 	mpd.wbc = wbc;
 	ext4_io_submit_init(&mpd.io_submit, wbc);
@@ -4015,6 +4015,7 @@ int ext4_inode_attach_jinode(struct inod
 		jbd2_free_inode(jinode);
 	return 0;
 }
+EXPORT_SYMBOL(ext4_inode_attach_jinode);
 
 /*
  * ext4_truncate()
--- l1/fs/ext4/ext4.h	2017-01-24 04:21:28.989470587 +0800
+++ l2/fs/ext4/ext4.h	2017-01-24 04:32:45.589800986 +0800
@@ -2462,7 +2462,8 @@ extern int ext4_group_add_blocks(handle_
 extern int ext4_trim_fs(struct super_block *, struct fstrim_range *);
 
 /* inode.c */
+#define HAVE_LDISKFS_INFO_JINODE
 struct buffer_head *ext4_getblk(handle_t *, struct inode *,
 						ext4_lblk_t, int, int *);
 struct buffer_head *ext4_bread(handle_t *, struct inode *,
 						ext4_lblk_t, int, int *);
