Index: linux-5.14.0-427.24.1.el9_4/fs/ext4/ext4.h
===================================================================
--- linux-5.14.0-427.24.1.el9_4.orig/fs/ext4/ext4.h
+++ linux-5.14.0-427.24.1.el9_4/fs/ext4/ext4.h
@@ -3176,7 +3176,8 @@ typedef enum {
 	EXT4_IGET_SPECIAL =	0x0001, /* OK to iget a system inode */
 	EXT4_IGET_HANDLE = 	0x0002,	/* Inode # is from a handle */
 	EXT4_IGET_BAD =		0x0004, /* Allow to iget a bad inode */
-	EXT4_IGET_EA_INODE =	0x0008	/* Inode should contain an EA value */
+	EXT4_IGET_EA_INODE =	0x0008,	/* Inode should contain an EA value */
+	EXT4_IGET_NO_CHECKS =	0x0010	/* Allow iget without checking */
 } ext4_iget_flags;
 
 extern struct inode *__ext4_iget(struct super_block *sb, unsigned long ino,
Index: linux-5.14.0-427.24.1.el9_4/fs/ext4/inode.c
===================================================================
--- linux-5.14.0-427.24.1.el9_4.orig/fs/ext4/inode.c
+++ linux-5.14.0-427.24.1.el9_4/fs/ext4/inode.c
@@ -4834,6 +4834,8 @@ static inline void ext4_inode_set_iversi
 static const char *check_igot_inode(struct inode *inode, ext4_iget_flags flags)
 
 {
+	if (flags & EXT4_IGET_NO_CHECKS)
+		return NULL;
 	if (flags & EXT4_IGET_EA_INODE) {
 		if (!(EXT4_I(inode)->i_flags & EXT4_EA_INODE_FL))
 			return "missing EA_INODE flag";
