Index: linux-2.6.9-42.0.10.EL_lustre.1.4.10/fs/ext3/super.c
===================================================================
--- linux-2.6.9-42.0.10.EL_lustre.1.4.10.orig/fs/ext3/super.c	2007-05-16 08:46:24.000000000 +0200
+++ linux-2.6.9-42.0.10.EL_lustre.1.4.10/fs/ext3/super.c	2007-05-16 08:48:58.000000000 +0200
@@ -123,6 +123,8 @@ void ext3_journal_abort_handle(const cha
 	journal_abort_handle(handle);
 }
 
+EXPORT_SYMBOL(ext3_journal_abort_handle);
+
 /* Deal with the reporting of failure conditions on a filesystem such as
  * inconsistencies detected or read IO failures.
  *
@@ -2064,6 +2066,8 @@ int ext3_force_commit(struct super_block
 	return ret;
 }
 
+EXPORT_SYMBOL(ext3_force_commit);
+
 /*
  * Ext3 always journals updates to the superblock itself, so we don't
  * have to propagate any other updates to the superblock on disk at this
@@ -2586,6 +2590,12 @@ int ext3_map_inode_page(struct inode *in
                         unsigned long *blocks, int *created, int create);
 EXPORT_SYMBOL(ext3_map_inode_page);
 
+EXPORT_SYMBOL(ext3_xattr_get);
+EXPORT_SYMBOL(ext3_xattr_set_handle);
+EXPORT_SYMBOL(ext3_bread);
+EXPORT_SYMBOL(ext3_journal_start_sb);
+EXPORT_SYMBOL(__ext3_journal_stop);
+
 MODULE_AUTHOR("Remy Card, Stephen Tweedie, Andrew Morton, Andreas Dilger, Theodore Ts'o and others");
 MODULE_DESCRIPTION("Second Extended Filesystem with journaling extensions");
 MODULE_LICENSE("GPL");
