Index: linux-stage/fs/ext3/super.c
===================================================================
--- linux-stage.orig/fs/ext3/super.c	2005-02-25 14:50:46.077845320 +0200
+++ linux-stage/fs/ext3/super.c	2005-02-25 14:51:32.241827328 +0200
@@ -123,6 +123,8 @@
 	journal_abort_handle(handle);
 }
 
+EXPORT_SYMBOL(ext3_journal_abort_handle);
+
 /* Deal with the reporting of failure conditions on a filesystem such as
  * inconsistencies detected or read IO failures.
  *
@@ -2002,6 +2004,8 @@
 	return ret;
 }
 
+EXPORT_SYMBOL(ext3_force_commit);
+
 /*
  * Ext3 always journals updates to the superblock itself, so we don't
  * have to propagate any other updates to the superblock on disk at this
@@ -2433,6 +2437,10 @@
                         unsigned long *blocks, int *created, int create);
 EXPORT_SYMBOL(ext3_map_inode_page);
 
+EXPORT_SYMBOL(ext3_xattr_get);
+EXPORT_SYMBOL(ext3_xattr_set_handle);
+EXPORT_SYMBOL(ext3_bread);
+
 MODULE_AUTHOR("Remy Card, Stephen Tweedie, Andrew Morton, Andreas Dilger, Theodore Ts'o and others");
 MODULE_DESCRIPTION("Second Extended Filesystem with journaling extensions");
 MODULE_LICENSE("GPL");
