Index: linux-2.6.5-sles9/include/linux/ext3_fs_sb.h
===================================================================
--- linux-2.6.5-sles9.orig/include/linux/ext3_fs_sb.h	2004-11-03 08:36:51.000000000 +0300
+++ linux-2.6.5-sles9/include/linux/ext3_fs_sb.h	2004-11-09 02:20:51.598024096 +0300
@@ -19,9 +19,12 @@
 #ifdef __KERNEL__
 #include <linux/timer.h>
 #include <linux/wait.h>
+#ifndef EXT_INCLUDE
+#define EXT_INCLUDE
 #include <linux/blockgroup_lock.h>
 #include <linux/percpu_counter.h>
 #endif
+#endif
 #include <linux/rbtree.h>
 
 /*
