Index: linux-2.6.0/include/linux/ext3_fs_sb.h
===================================================================
--- linux-2.6.0.orig/include/linux/ext3_fs_sb.h	2003-06-24 18:05:26.000000000 +0400
+++ linux-2.6.0/include/linux/ext3_fs_sb.h	2003-12-31 01:09:26.000000000 +0300
@@ -19,9 +19,12 @@
 #ifdef __KERNEL__
 #include <linux/timer.h>
 #include <linux/wait.h>
+#ifndef EXT_INCLUDE
+#define EXT_INCLUDE
 #include <linux/blockgroup_lock.h>
 #include <linux/percpu_counter.h>
 #endif
+#endif
 
 /*
  * third extended-fs super-block data in memory
