Index: linux-stage/fs/ext3/ialloc.c
===================================================================
--- linux-stage.orig/fs/ext3/ialloc.c	2005-06-26 10:59:43.048185981 +0200
+++ linux-stage/fs/ext3/ialloc.c	2005-06-26 11:01:21.317716027 +0200
@@ -775,7 +775,6 @@
 		if (!gdp)
 			continue;
 		desc_count += le16_to_cpu(gdp->bg_free_inodes_count);
-		cond_resched();
 	}
 	return desc_count;
 #endif
Index: linux-stage/fs/ext3/super.c
===================================================================
--- linux-stage.orig/fs/ext3/super.c	2005-06-26 10:59:43.205412542 +0200
+++ linux-stage/fs/ext3/super.c	2005-06-26 11:02:29.599941754 +0200
@@ -2236,11 +2232,9 @@
 		 * block group descriptors.  If the sparse superblocks
 		 * feature is turned on, then not all groups have this.
 		 */
-		for (i = 0; i < ngroups; i++) {
+		for (i = 0; i < ngroups; i++)
 			overhead += ext3_bg_has_super(sb, i) +
 				ext3_bg_num_gdb(sb, i);
-			cond_resched();
-		}
 
 		/*
 		 * Every block group has an inode bitmap, a block
