Index: lustre+chaos/lustre/ldiskfs/namei.c
===================================================================
--- linux-2.6.5-7.283-full.orig/fs/ext3/namei.c
+++ linux-2.6.5-7.283-full/fs/ext3/namei.c
@@ -2184,8 +2184,8 @@
 	 * Return -ENOENT if we've raced with unlink and i_nlink is 0.  Doing
 	 * otherwise has the potential to corrupt the orphan inode list.
 	 */
-	if (inode->i_nlink == 0)
-		return -ENOENT;
+	//if (inode->i_nlink == 0)
+	//	return -ENOENT;
 
 retry:
 	handle = ext3_journal_start(dir, EXT3_DATA_TRANS_BLOCKS +
