Index: linux-2.6.9-67.0.22/fs/ext3/ioctl.c
===================================================================
--- linux-2.6.9-67.0.22.orig/fs/ext3/ioctl.c
+++ linux-2.6.9-67.0.22/fs/ext3/ioctl.c
@@ -157,7 +157,7 @@ int ioctl_fiemap(struct inode *inode, st
 		       return -EFAULT;
 
 	if (fieinfo.fi_flags & FIEMAP_FLAG_SYNC)
-		filemap_write_and_wait(inode->i_mapping);
+		ext3_sync_file(filp, filp->f_dentry, 1);
 
 	error = ext3_fiemap(inode, &fieinfo, fiemap.fm_start, len);
 	fiemap.fm_flags = fieinfo.fi_flags;
