Index: linux-2.6.16.60-0.33_org/fs/ext3/hash.c
===================================================================
--- linux-2.6.16.60-0.33_org/fs/ext3/hash.c	2009-07-01 18:16:50.000000000 +0530
+++ linux-2.6.16.60-0.33_new/fs/ext3/hash.c	2009-07-01 18:26:58.000000000 +0530
@@ -8,7 +8,7 @@
  * This file may be redistributed under the terms of the GNU Public
  * License.
  */
-
+#include <linux/module.h>
 #include <linux/fs.h>
 #include <linux/jbd.h>
 #include <linux/sched.h>
@@ -173,3 +173,4 @@ int ext3fs_dirhash(const char *name, int
 	hinfo->minor_hash = minor_hash;
 	return 0;
 }
+EXPORT_SYMBOL(ext3fs_dirhash);
Index: linux-2.6.16.60-0.33_org/fs/ext3/namei.c
===================================================================
--- linux-2.6.16.60-0.33_org/fs/ext3/namei.c	2009-07-01 18:16:50.000000000 +0530
+++ linux-2.6.16.60-0.33_new/fs/ext3/namei.c	2009-07-01 18:24:49.000000000 +0530
@@ -75,6 +75,7 @@ struct buffer_head *ext3_append(handle_t
 
 	return bh;
 }
+EXPORT_SYMBOL(ext3_append);
 
 #ifndef assert
 #define assert(test) J_ASSERT(test)
Index: linux-2.6.16.60-0.33_org/fs/ext3/super.c
===================================================================
--- linux-2.6.16.60-0.33_org/fs/ext3/super.c	2009-07-01 18:16:50.000000000 +0530
+++ linux-2.6.16.60-0.33_new/fs/ext3/super.c	2009-07-01 18:24:27.000000000 +0530
@@ -260,6 +260,7 @@ void __ext3_std_error (struct super_bloc
 
 	ext3_handle_error(sb);
 }
+EXPORT_SYMBOL(__ext3_std_error);
 
 /*
  * ext3_abort is a much stronger failure handler than ext3_error.  The
Index: linux-2.6.16.60-0.33_org/include/linux/ext3_fs.h
===================================================================
--- linux-2.6.16.60-0.33_org/include/linux/ext3_fs.h	2009-07-01 18:16:50.000000000 +0530
+++ linux-2.6.16.60-0.33_new/include/linux/ext3_fs.h	2009-07-01 18:22:09.000000000 +0530
@@ -1055,6 +1055,8 @@ extern void ext3_abort (struct super_blo
 extern void ext3_warning (struct super_block *, const char *, const char *, ...)
 	__attribute__ ((format (printf, 3, 4)));
 extern void ext3_update_dynamic_rev (struct super_block *sb);
+extern void __ext3_std_error (struct super_block * sb, const char * function,
+				int errno);
 
 #define ext3_std_error(sb, errno)				\
 do {								\
@@ -1076,6 +1078,8 @@ extern struct file_operations ext3_file_
 /* namei.c */
 extern struct inode_operations ext3_dir_inode_operations;
 extern struct inode_operations ext3_special_inode_operations;
+extern struct buffer_head *ext3_append(handle_t *handle, struct inode *inode,
+					u32 *block, int *err);
 
 /* symlink.c */
 extern struct inode_operations ext3_symlink_inode_operations;

