diff -rupN linux-2.6.18-53.1.14.org/fs/ext3/namei.c linux-2.6.18-53.1.14.new/fs/ext3/namei.c
--- linux-2.6.18-53.1.14.org/fs/ext3/namei.c	2008-10-21 16:08:28.000000000 +0530
+++ linux-2.6.18-53.1.14.new/fs/ext3/namei.c	2008-10-21 16:09:25.000000000 +0530
@@ -1612,7 +1612,7 @@ cleanup:
  * ext3_delete_entry deletes a directory entry by merging it with the
  * previous entry
  */
-static int ext3_delete_entry (handle_t *handle, 
+static int ext3_delete_entry (handle_t *handle,
 			      struct inode * dir,
 			      struct ext3_dir_entry_2 * de_del,
 			      struct buffer_head * bh)
