diff -rupN linux-2.6.18-128.1.6_1/fs/ext3/Makefile linux-2.6.18-128.1.6_2/fs/ext3/Makefile
--- linux-2.6.18-128.1.6_1/fs/ext3/Makefile	2009-08-13 19:19:54.000000000 +0530
+++ linux-2.6.18-128.1.6_2/fs/ext3/Makefile	2009-08-13 19:20:30.000000000 +0530
@@ -5,7 +5,8 @@
 obj-$(CONFIG_EXT3_FS) += ext3.o
 
 ext3-y	:= balloc.o bitmap.o dir.o file.o fsync.o ialloc.o inode.o \
-	   ioctl.o namei.o super.o symlink.o hash.o resize.o extents.o mballoc.o
+	   ioctl.o namei.o super.o symlink.o hash.o resize.o extents.o \
+	   mballoc.o dynlocks.o
 
 ext3-$(CONFIG_EXT3_FS_XATTR)	 += xattr.o xattr_user.o xattr_trusted.o
 ext3-$(CONFIG_EXT3_FS_POSIX_ACL) += acl.o
diff -rupN linux-2.6.18-128.1.6_1/fs/ext3/super.c linux-2.6.18-128.1.6_2/fs/ext3/super.c
--- linux-2.6.18-128.1.6_1/fs/ext3/super.c	2009-08-13 19:19:54.000000000 +0530
+++ linux-2.6.18-128.1.6_2/fs/ext3/super.c	2009-08-13 19:23:23.000000000 +0530
@@ -3529,6 +3530,7 @@ static int __init init_ext3_fs(void)
 	err = init_inodecache();
 	if (err)
 		goto out1;
+	dynlock_cache_init();
         err = register_filesystem(&ext3_fs_type);
 	if (err)
 		goto out;
@@ -3546,6 +3548,7 @@ out1:
 static void __exit exit_ext3_fs(void)
 {
 	unregister_filesystem(&ext3_fs_type);
+	dynlock_cache_exit();
 	destroy_inodecache();
 	exit_ext3_xattr();
 	exit_ext3_proc();

