--- linux-src.org/fs/ext3/ext3_jbd.c
+++ linux-src/fs/ext3/ext3_jbd.c
@@ -2,6 +2,7 @@
  * Interface between ext3 and JBD
  */
 
+#include <linux/module.h>
 #include <linux/ext3_jbd.h>
 
 int __ext3_journal_get_undo_access(const char *where, handle_t *handle,
@@ -21,6 +21,7 @@ int __ext3_journal_get_write_access(cons
 		ext3_journal_abort_handle(where, __FUNCTION__, bh, handle,err);
 	return err;
 }
+EXPORT_SYMBOL(__ext3_journal_get_write_access);
 
 int __ext3_journal_forget(const char *where, handle_t *handle,
 				struct buffer_head *bh)
@@ -57,3 +58,5 @@ int __ext3_journal_dirty_metadata(const 
 		ext3_journal_abort_handle(where, __FUNCTION__, bh, handle,err);
 	return err;
 }
+
+EXPORT_SYMBOL(__ext3_journal_dirty_metadata);
