---
 fs/ext3/super.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- linux-2.6.16.21-0.8.orig/fs/ext3/super.c
+++ linux-2.6.16.21-0.8/fs/ext3/super.c
@@ -1425,7 +1425,7 @@ static int ext3_fill_super (struct super
 	sbi->s_resgid = le16_to_cpu(es->s_def_resgid);
 
 	/* enable barriers by default */
-	set_opt(sbi->s_mount_opt, BARRIER);
+	/* set_opt(sbi->s_mount_opt, BARRIER); */
 	set_opt(sbi->s_mount_opt, RESERVATION);
 
 	if (!parse_options ((char *) data, sb, &journal_inum, &journal_devnum,
