Index: linux-stage/include/linux/ext3_fs.h
===================================================================
--- linux-stage.orig/include/linux/ext3_fs.h	2004-04-02 16:43:37.000000000 -0500
+++ linux-stage/include/linux/ext3_fs.h	2004-04-02 16:43:37.000000000 -0500
@@ -331,12 +331,13 @@
 #define EXT3_MOUNT_IOPEN_NOPRIV		0x100000/* Make iopen world-readable */
 
 /* Compatibility, for having both ext2_fs.h and ext3_fs.h included at once */
-#ifndef _LINUX_EXT2_FS_H
+#ifndef clear_opt
 #define clear_opt(o, opt)		o &= ~EXT3_MOUNT_##opt
 #define set_opt(o, opt)			o |= EXT3_MOUNT_##opt
 #define test_opt(sb, opt)		(EXT3_SB(sb)->s_mount_opt & \
 					 EXT3_MOUNT_##opt)
-#else
+#endif
+#ifndef EXT2_MOUNT_NOLOAD
 #define EXT2_MOUNT_NOLOAD		EXT3_MOUNT_NOLOAD
 #define EXT2_MOUNT_ABORT		EXT3_MOUNT_ABORT
 #define EXT2_MOUNT_DATA_FLAGS		EXT3_MOUNT_DATA_FLAGS
