Index: linux-stage/fs/ext4/Makefile
===================================================================
--- linux-stage.orig/fs/ext4/Makefile
+++ linux-stage/fs/ext4/Makefile
@@ -7,7 +7,7 @@ obj-$(CONFIG_EXT4_FS) += ext4.o
 ext4-y	:= balloc.o bitmap.o dir.o file.o fsync.o ialloc.o inode.o \
 		ioctl.o namei.o super.o symlink.o hash.o resize.o extents.o \
 		ext4_jbd2.o migrate.o mballoc.o block_validity.o move_extent.o \
-		mmp.o
+		mmp.o dynlocks.o
 
 ext4-$(CONFIG_EXT4_FS_XATTR)		+= xattr.o xattr_user.o xattr_trusted.o
 ext4-$(CONFIG_EXT4_FS_POSIX_ACL)	+= acl.o
Index: linux-stage/fs/ext4/super.c
===================================================================
--- linux-stage.orig/fs/ext4/super.c
+++ linux-stage/fs/ext4/super.c
@@ -4159,6 +4159,7 @@ static int __init init_ext4_fs(void)
 	err = init_inodecache();
 	if (err)
 		goto out1;
+	dynlock_cache_init();
 	err = register_filesystem(&ext4_fs_type);
 	if (err)
 		goto out;
@@ -4195,6 +4196,7 @@ static void __exit exit_ext4_fs(void)
 	unregister_filesystem(&ext4dev_fs_type);
 #endif
 	destroy_inodecache();
+	dynlock_cache_exit();
 	exit_ext4_xattr();
 	exit_ext4_mballoc();
 	__free_page(ext4_zero_page);
