diff -rupN linux-2.6.18-128.1.6_org/fs/ext4/Makefile linux-2.6.18-128.1.6_org1/fs/ext4/Makefile
--- linux-2.6.18-128.1.6_org/fs/ext4/Makefile	2009-07-08 18:00:38.000000000 +0530
+++ linux-2.6.18-128.1.6_org1/fs/ext4/Makefile	2009-07-10 18:24:52.000000000 +0530
@@ -6,7 +6,7 @@ obj-$(CONFIG_EXT4DEV_FS) += ext4dev.o
 
 ext4dev-y	:= balloc.o bitmap.o dir.o file.o fsync.o ialloc.o inode.o iopen.o \
 		   ioctl.o namei.o super.o symlink.o hash.o resize.o extents.o \
-		   ext4_jbd2.o migrate.o mballoc.o
+		   ext4_jbd2.o migrate.o mballoc.o dynlocks.o
 
 ext4dev-$(CONFIG_EXT4DEV_FS_XATTR)	+= xattr.o xattr_user.o xattr_trusted.o
 ext4dev-$(CONFIG_EXT4DEV_FS_POSIX_ACL)	+= acl.o
diff -rupN linux-2.6.18-128.1.6_org/fs/ext4/super.c linux-2.6.18-128.1.6_org1/fs/ext4/super.c
--- linux-2.6.18-128.1.6_org/fs/ext4/super.c	2009-07-08 18:00:38.000000000 +0530
+++ linux-2.6.18-128.1.6_org1/fs/ext4/super.c	2009-07-10 18:54:27.000000000 +0530
@@ -3890,6 +3890,8 @@ static struct file_system_type ext4_fs_t
 #endif
 };
 
+extern void dynlock_cache_init(void);
+
 static int __init init_ext4_fs(void)
 {
 	int err;
@@ -3910,6 +3912,7 @@ static int __init init_ext4_fs(void)
 	err = init_inodecache();
 	if (err)
 		goto out1;
+	dynlock_cache_init();
 	err = register_filesystem(&ext4_fs_type);
 	if (err)
 		goto out;
@@ -3925,9 +3928,12 @@ out3:
 	return err;
 }
 
+extern void dynlock_cache_exit(void);
+
 static void __exit exit_ext4_fs(void)
 {
 	unregister_filesystem(&ext4_fs_type);
+	dynlock_cache_exit();
 	destroy_inodecache();
 	exit_ext4_xattr();
 	exit_ext4_mballoc();
