Index: linux-2.6.18.i386/fs/ext4/ext4_i.h
===================================================================
--- linux-2.6.18.i386.orig/fs/ext4/ext4_i.h
+++ linux-2.6.18.i386/fs/ext4/ext4_i.h
@@ -162,6 +162,8 @@ struct ext4_inode_info {
	__u16 i_extra_isize;

	spinlock_t i_block_reservation_lock;
+
+	void *i_filterdata;
 };
 
 #endif	/* _EXT4_I */
Index: linux-2.6.18.i386/fs/ext4/super.c
===================================================================
--- linux-2.6.18.i386.orig/fs/ext4/super.c
+++ linux-2.6.18.i386/fs/ext4/super.c
@@ -574,6 +574,7 @@ static struct inode *ext4_alloc_inode(st
	ei->i_allocated_meta_blocks = 0;
	ei->i_delalloc_reserved_flag = 0;
	spin_lock_init(&(ei->i_block_reservation_lock));
+	ei->i_filterdata = NULL;
 	return &ei->vfs_inode;
 }
 
