Index: linux-stage/fs/ext4/super.c
===================================================================
--- linux-stage.orig/fs/ext4/super.c
+++ linux-stage/fs/ext4/super.c
@@ -2564,10 +2564,6 @@ static int ext4_fill_super(struct super_
 	*/
 	if (EXT4_HAS_INCOMPAT_FEATURE(sb, EXT4_FEATURE_INCOMPAT_EXTENTS))
 		set_opt(sbi->s_mount_opt, EXTENTS);
-	else
-		ext4_warning(sb,
-			"extents feature not enabled on this filesystem, "
-			"use tune2fs.");
 
 	/*
 	 * enable delayed allocation by default
